{
 "cells": [
  {
   "cell_type": "raw",
   "id": "e15ca787-fd40-4888-94e7-b78a721b05b9",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Microbiome Sequence Analysis\"\n",
    "subtitle: \"Community composition from 16S Sequence Data of Montipora capitata corals exposed to heat and plasticizers\"\n",
    "author: \"Sarah Tanja\"\n",
    "format: html\n",
    "editor: visual\n",
    "toc: true\n",
    "toc-title: Contents <i class=\"bi bi-bookmark-heart\"></i>\n",
    "toc-depth: 5\n",
    "toc-location: left\n",
    "reference-location: margin\n",
    "citation-location: margin\n",
    "bibliography: ./references/qiime2.bib\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "122fda22-a268-433a-b619-ec18f7946413",
   "metadata": {},
   "source": [
    "# Amplicon Sequence Data Analysis with QIIME 2\n",
    "\n",
    "Adding ```!``` before the command tells the notebook this is a bash command, rather than python.\n",
    "\n",
    "### Data download\n",
    "I downloaded two .zip folders from a Dropbox link; one had raw data files and the other had analysis pipeline files generated by [Mr. DNA Lab](https://www.mrdnalab.com/) Molecular Research. I unzipped the folders and uploaded the Mr. DNA analysis pipeline files into the `coral-pae-temp/analysis/microbiome/mrdna` directory, and the `sample-metadata.tsv` and `demux` folder into `coral-pae-temp/analysis/microbiome/rawdata` directory.  \n",
    "\n",
    "### Data type\n",
    "\n",
    "Here I am working with the raw data files located in `coral-pae-temp/analysis/microbiome/rawdata/demux`. In the `demux` folder are two `fastq.gz` files for each of the 22 samples, one for the forward read and one for the reverse read. \n",
    "\n",
    "The `fastq.gz` file name includes the sample identifier and should look like `4.Ea_S1_L001_R1_001.fastq.gz`. \n",
    "The underscore-separated fields in this file name are:\n",
    "\n",
    "1.  the sample identifier,\n",
    "\n",
    "2.  the barcode sequence or a barcode identifier,\n",
    "\n",
    "3.  the lane number,\n",
    "\n",
    "4.  the direction of the read (i.e. R1 or R2, because these are paired-end reads), and\n",
    "\n",
    "5.  the set number.\n",
    "   \n",
    "\n",
    "The `fastq.gz` files are **Demultiplexed** (aka **Demuxed**) sequences that still have the forward and reverse primers in the sequences.\n",
    "\n",
    "-   The Raw Data is **demultiplexed**\n",
    "\n",
    "-   A R1 and R2 fastq.gz file has been generated for each individual sample\n",
    "\n",
    "-   All forward reads are binned into the R1 fastq.gz files\n",
    "\n",
    "-   All reverse reads are binned into the R2 fastq.gz files\n",
    "\n",
    "-   Other than demultiplexing; you can consider the Raw Data on BaseSpace as untouched (**The Forward and Reverse Primer Sequences have not been removed**)\n",
    "\n",
    "### Data process\n",
    "Here I follow the QIIME2 [Casava 1.8 paired-end demultiplexed fastq](https://docs.qiime2.org/2023.5/tutorials/importing/#:~:text=Casava%201.8%20paired%2Dend%20demultiplexed%20fastq) tutorial example on importing data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7997d9-10ca-4323-85e2-676ee96f60b5",
   "metadata": {},
   "source": [
    "## Python 3 API import qiime plugins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a593cfd4-fcb7-4a42-8440-457f99e65015",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18dc8856-97ef-42d6-9f4c-3cdd23ea0f83",
   "metadata": {},
   "source": [
    "## Take a Look at the Metadata\n",
    "Make a table of the metadata.\n",
    "Here I added columns 'Pae', 'Temp', 'PeaTemp', 'Colony', and 'Tank' to the original `sample-metadata.tsv` file provided to me by Mr. DNA and renamed it `sample-metadata-verbose.tsv`\n",
    "This was a bit of a process... I had to:\n",
    "upload the `sample-metadata.tsv` to Excel\n",
    "edit the metadata by adding the above columns and values\n",
    "save it as a csv\n",
    "open it in a text editor\n",
    "search for all ',' commas, and find&replace them with 'TAB' symbols\n",
    "save as a tab separated file `.tsv`\n",
    "upload it back into the `coral-pae-temp/analysis/microbiome/rawdata` folder\n",
    "\n",
    ":::{.callout-important}\n",
    "At first I had named the new columns 'pae', 'temp', etc. with lower case... for some reason this was a problem later on and the interactive emperor plots wouldn't recognize the new columns. When I changed the column names to CamelCase to match the others, it worked. The `qiime2` docs indicate that metadata formatted with an Identifier Column such as `#Sample ID` is [case-sensitive](https://docs.qiime2.org/2023.5/tutorials/metadata/#metadata-formatting-requirements:~:text=feature%2Did-,Case%2Dsensitive,-(these%20are%20mostly)\n",
    ":::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c420fcd4-0a0c-4fa7-a36f-4a1d4069b4ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/metadata-verbose.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/metadata-verbose.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd415645-5a55-4d00-8dfd-e2d2246a3c3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/357ae5fb-826d-403e-969d-ec023b3d7f34')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 357ae5fb-826d-403e-969d-ec023b3d7f34>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/metadata-verbose.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a04668c-4e70-4051-a3df-9374bf7751d1",
   "metadata": {},
   "source": [
    "# Import Sequences into QIIME2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e670c1b-4199-4cdf-9fa3-27c433bfcf6c",
   "metadata": {},
   "source": [
    "To use sequencing data in QIIME2, we first need to turn the FASTQ files containing our data into QIIME artifacts.\n",
    "\n",
    "What the QIIME2 pipeline will do:\r\n",
    "![our workflow](https://github.com/Gibbons-Lab/isb_course_2023/raw/main/docs/16S/assets/steps.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c48706f-b3df-430e-ae9c-e35005f44658",
   "metadata": {},
   "source": [
    "!qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path ../rawdata/demux \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path ../output/demux-paired-end.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17971af4-e264-4f18-b33a-3e774fdf7295",
   "metadata": {},
   "source": [
    "The `demux-paired-end.qza` artifact contains raw, demultiplexed sequences that still have forward and reverse primers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cc3897b-7598-406d-8813-da555c48ebb8",
   "metadata": {},
   "source": [
    "## Trim primers from paired-end sequences using `cutadapt`\n",
    "\n",
    "> \"The PCR primers (F515/R806) were developed against the V4 region of the 16S rRNA, which we determined would yield optimal community clustering with reads of this length using a procedure similar to that of ref. 15. [For reference, this primer pair amplifies the region 533–786 in the Escherichia coli strain 83972 sequence (greengenes accession no. prokMSA_id:470367).] The reverse PCR primer is barcoded with a 12-base error correcting Golay code to facilitate multiplexing of up to ≈1,500 samples per lane, and both PCR primers contain sequencer adapter regions.\" - (Caporasco et al. 2011)\n",
    "\n",
    "Caporaso, J. G., Lauber, C. L., Walters, W. A., Berg-Lyons, D., Lozupone, C. A., Turnbaugh, P. J., Fierer, N., & Knight, R. (2011). Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample. Proceedings of the National Academy of Sciences, 108(supplement_1), 4516–4522. https://doi.org/10.1073/pnas.1000080107\n",
    "\n",
    "> \"The V4 variable region of the 16S rRNA gene was amplified using the 515F (5′-­GTGCCAGCMGCCGCGGTAA-­3′) and 806R (5′-­GGACTACHVGGGTWTCTAAT-­3′) primer set (Caporaso et al. 2011). ” - (Brown et al. 2021)\n",
    "\n",
    "Brown, Tanya, Dylan Sonett, Jesse R. Zaneveld, and Jacqueline L. Padilla-Gamiño. 2021. “Characterization of the Microbiome and Immune Response in Corals with Chronic Montipora White Syndrome.” Molecular Ecology 30 (11): 2591–2606. https://doi.org/10.1111/mec.15899."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "967c62c9-a86a-48f6-9d5f-6d6a5501e1ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved SampleData[PairedEndSequencesWithQuality] to: ../output/demux-trimmed.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime cutadapt trim-paired \\\n",
    "  --i-demultiplexed-sequences ../output/demux-paired-end.qza \\\n",
    "  --p-cores 4 \\\n",
    "  --p-front-f GTGYCAGCMGCCGCGGTAA \\\n",
    "  --p-front-r GGACTACNVGGGTWTCTAAT \\\n",
    "  --o-trimmed-sequences ../output/demux-trimmed.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00901ad",
   "metadata": {},
   "source": [
    "## Visualize trimmed & demultiplexed sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69b6a077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/demux-trimmed-summary.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime demux summarize \\\n",
    "  --i-data ../output/demux-trimmed.qza \\\n",
    "  --o-visualization ../output/demux-trimmed-summary.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dc0de45-a09c-4363-a838-5d1638b584ac",
   "metadata": {},
   "source": [
    "We are looking for at least 10,000 reads per sample. We have a minimunm of 205, 508 reads, so we are well above the minimal threshold for 'enough' reads to work with. Some of these reads will be removed in the downstream processing when we 'denoise' using the DADA2 algorithm to remove low quality reads, amplicon sequence variants, and chimeras."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da27f217-31b2-4cb5-8edb-6fd6bd70f48f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/2239e507-40ef-48de-b79d-77d9ae53efe3')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 2239e507-40ef-48de-b79d-77d9ae53efe3>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/demux-trimmed-summary.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873d9098-a910-4845-ade0-8b55c7622239",
   "metadata": {},
   "source": [
    "## [Denoise with DADA2](https://docs.qiime2.org/2023.5/tutorials/moving-pictures/#sequence-quality-control-and-feature-table-construction:~:text=with%20QIIME%201.-,Option%201%3A%20DADA2%C2%B6,-DADA2%20is%20a)\n",
    "\n",
    "[DADA2](https://pubmed.ncbi.nlm.nih.gov/27214047/) is a pipeline for detecting and correcting (where possible) Illumina amplicon sequence data. \n",
    "\n",
    "As implemented in the q2-dada2 plugin, this quality control process will additionally filter any phiX reads (commonly present in marker gene Illumina sequence data) that are identified in the sequencing data, and will filter chimeric sequences.  \n",
    "The dada2 denoise-paired method requires four parameters that are used in quality filtering:  \n",
    "\n",
    "    `--p-trim-left-f m`, which trims off the first m bases of each sequence in the forward reads\n",
    "    `--p-trim-left-r n`, which trims off the first m bases of each sequence in the reverse reads\n",
    "    `--p-trunc-len-f o`, which truncates each sequence at position o in the forward reads\n",
    "    `--p-trunc-len-f p`, which truncates each sequence at position p in the reverse reads  \n",
    "    \n",
    "This allows the user to remove low quality regions of the sequences.  \n",
    "\n",
    "What is a 'good' quality score?  \n",
    "\n",
    "In QIIME 2's interactive quality plots, the quality scores typically range from 0 to 40. Quality scores reflect the accuracy of base calls in sequencing data, with higher scores indicating higher accuracy. The most common quality score scale used in modern sequencing technologies is the Phred scale.\r\n",
    "\r\n",
    "In the Phred sceal - \r\n",
    "\r\n",
    "A quality score of 10 corresponds to a 1 in 10 chance of an incorrect base call (90% aacc  y).\r\n",
    "A quality score of 20 corresponds to a 1 in 100 chance of an incorrect base call (99%r ua  cy).\r\n",
    "A quality score of 30 corresponds to a 1 in 1000 chance of an incorrect base call  (9curac9.9%\n",
    "accur\n",
    "A \"good\" quality score in this context depends on your specific analysis goals and the sequencing platform you're u\n",
    "sing. However, many researchers consider quality s3b and aboveove 20 to be generally acceptable for downstreaanal**ysis. Scores above 30 are often seen as very ha qu*. In theew the Interactive Quality Plot tab in the `demux-trimmed-summary.qzv` file that was generated by `qiime demux summarlots, we see that the quality scores of the bases are high, between a score of 11 in the lowest 2nd percentile and a score of 37 in the bottom 25th percentile and hitter.  \n",
    "\n",
    "[Denoising tips from Greg Caporasco](https://docs.qiime2.org/jupyterbooks/cancer-microbiome-intervention-tutorial/020-tutorial-upstream/040-denoising.html) using `qiime dada2 denoise-paired`\n",
    "\n",
    " - Forward- 230\n",
    " - Reve31\n",
    "\n",
    "This step will take 10-15 minutes!t step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fc5e87b8-0b3b-46ab-b1e2-4291fb1ed11b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: ../output/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: ../output/rep-seqs.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: ../output/denoising-stats.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs ../output/demux-trimmed.qza \\\n",
    "  --p-trim-left-f 0 \\\n",
    "  --p-trim-left-r 0 \\\n",
    "  --p-trunc-len-f 230 \\\n",
    "  --p-trunc-len-r 231 \\\n",
    "  --p-n-threads 20 \\\n",
    "  --o-table ../output/table.qza \\\n",
    "  --o-representative-sequences ../output/rep-seqs.qza \\\n",
    "  --o-denoising-stats ../output/denoising-stats.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9f6d43cc-5721-4885-a363-4bd948e7fc90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada.R --input_directory /tmp/tmpwef8c0hn/forward --input_directory_reverse /tmp/tmpwef8c0hn/reverse --output_path /tmp/tmpwef8c0hn/output.tsv.biom --output_track /tmp/tmpwef8c0hn/track.tsv --filtered_directory /tmp/tmpwef8c0hn/filt_f --filtered_directory_reverse /tmp/tmpwef8c0hn/filt_r --truncation_length 230 --truncation_length_reverse 231 --trim_left 0 --trim_left_reverse 0 --max_expected_errors 2.0 --max_expected_errors_reverse 2.0 --truncation_quality_score 2 --min_overlap 12 --pooling_method independent --chimera_method consensus --min_parental_fold 1.0 --allow_one_off False --num_threads 20 --learn_min_reads 1000000\n",
      "\n",
      "R version 4.2.3 (2023-03-15) \n",
      "Loading required package: Rcpp\n",
      "\u001b[?25hDADA2: 1.26.0 / Rcpp: 1.0.10 / RcppParallel: 5.1.6 \n",
      "\u001b[?25h\u001b[?25h2) Filtering \u001b[?25h\u001b[?25h......................\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "\u001b[?25h\u001b[?25h3) Learning Error Rates\n",
      "\u001b[?25h240503870 total bases in 1045669 reads from 3 samples will be used for learning the error rates.\n",
      "241549539 total bases in 1045669 reads from 3 samples will be used for learning the error rates.\n",
      "\u001b[?25h3) Denoise samples ......................\n",
      "......................\n",
      "\u001b[?25h5) Remove chimeras (method = consensus)\n",
      "\u001b[?25h\u001b[?25h6) Report read numbers through the pipeline\n",
      "\u001b[?25h\u001b[?25h7) Write output\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[32mSaved FeatureTable[Frequency] to: ../output/dada/table.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[Sequence] to: ../output/dada/representative_sequences.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[DADA2Stats] to: ../output/dada/denoising_stats.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs ../output/demux-trimmed.qza \\\n",
    "  --p-trim-left-f 0 \\\n",
    "  --p-trim-left-r 0 \\\n",
    "  --p-trunc-len-f 230 \\\n",
    "  --p-trunc-len-r 231 \\\n",
    "  --p-n-threads 20 \\\n",
    "  --output-dir ../output/dada --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3ff89ca9-6ec3-4d24-962f-c3f21f8249d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/dada/denoising-stats.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file ../output/denoising-stats.qza \\\n",
    "  --o-visualization ../output/dada/denoising-stats.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "043972a4-d961-43d4-b008-622b0b5a94ab",
   "metadata": {},
   "source": [
    "### Visualize Denoising Stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17e0dc09-05b6-4251-a7c3-9e960f9807c8",
   "metadata": {},
   "source": [
    "One good way to tell if the identified ASVs are representative of the sample is to see how many reads were maintained throughout the pipeline. Here, the most common issues and solutions are:\r\n",
    "\r\n",
    "**Large fraction of reads is lost during merging (only paired-end)**\r\n",
    "\r\n",
    "![read overlap](https://gibbons-lab.github.io/isb_course_2023/16S/assets/read_overlap.png)\r\n",
    "\r\n",
    "In order to merge ASVs DADA2 uses an overlap of 12 bases between forward and reverse reads by default. Thus, your reads must allow for sufficient overlap *after* trimming. So if your amplified region is 450bp long and you have 2x250bp reads and you trim the last 30 bases of each read, truncating the length to 220bp, the total length of covered sequence is 2x220 = 440 which is shorter than 450bp so there will be no overlap. To solve this issue trim less of the reads or adjust the `--p-min-overlap` parameters to something lower (but not too low).\r\n",
    "\r\n",
    "<br>\r\n",
    "\r\n",
    "**Most of the reads are lost as chimeric**\r\n",
    "\r\n",
    "![read overlap](https://gibbons-lab.github.io/isb_course_2023/16S/assets/chimera.png)\r\n",
    "\r\n",
    "This is usually an experimental issue as chimeras are introduced during amplification. If you can adjust your PCR, try to run fewer cycles. Chimeras can also be introduced by incorrect merging. If your minimum overlap is too small ASVs may be merged randomly. Possible fixes are to increase the `--p-min-overlap` parameter or run the analysis on the forward reads only (in our empirical observations, chimeras are more likely to be introduced in the joined reads). *However, losing between 5-25% of your reads to chimeras is normal and does not require any adjustments.*\r\n",
    "\r\n",
    "\r\n",
    "Our denoising stats are contained in an artifact. To convert it to a visualization we can use `qiime metadata tabulate`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "62530108-5547-4a61-941e-ad81864dc0ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/ff8ce97e-8b05-49f7-8bfd-bc3db40d6fbd')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: ff8ce97e-8b05-49f7-8bfd-bc3db40d6fbd>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/dada/denoising-stats.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "586afdc0-6bde-4b61-a10e-886d1de3a839",
   "metadata": {},
   "source": [
    "## Summarize & tabulate the feature table\n",
    "After the quality filtering step completes, you’ll want to explore the resulting data. You can do this using the following two commands, which will create visual summaries of the data. The `feature-table summarize` command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics. The `feature-table tabulate-seqs` command will provide a mapping of feature IDs to sequences, and provide links to easily BLAST each sequence against the NCBI nt database.\n",
    "\n",
    "### feature-table summarize\n",
    "The feature-table summarize command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics. Feature tables in QIIME 2 represent the abundance of different biological features (such as bacterial taxa or OTUs) across samples.\n",
    " In this command:\n",
    "\n",
    "--i-table tab.qzal`es the input feature table in QIIME 2 artifact format (.qza file) that you want to sumri\r\n",
    "\r\n",
    "--o-visualizattable.qzv tfies the output visualization in QIIME 2 artifact format (.qzv file) that will con the summaes.\r\n",
    "\r\n",
    "--m-sample-metadata-file sampldata.tsve-cifies the metadata file (usually in tab-separated values format) that contains additional information about the samples in your featble  and m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "18f2f01e-f8cc-4feb-b0e6-d4316ef9a216",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/dada/table.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table summarize \\\n",
    "  --i-table ../output/dada/table.qza \\\n",
    "  --o-visualization ../output/dada/table.qzv \\\n",
    "  --m-sample-metadata-file ../rawdata/sample-metadata-verbose.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f74905c7-7813-44bd-a27d-0350d9866e36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/d2915e2b-7930-45c1-b6d5-f3396f17e767')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: d2915e2b-7930-45c1-b6d5-f3396f17e767>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/dada/table.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7144554a-99b0-4fbd-8a79-606433ad2fb7",
   "metadata": {},
   "source": [
    "### feature-table tabulate-seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "16801dde-0810-47ae-ba30-76867ef1be15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/dada/representative_sequences.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table tabulate-seqs \\\n",
    "  --i-data ../output/dada/representative_sequences.qza \\\n",
    "  --o-visualization ../output/dada/representative_sequences.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0e800ea2-04c3-4594-a96c-8689bf39b282",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/302a519e-a5e5-41e1-8c04-41c9d10765b4')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 302a519e-a5e5-41e1-8c04-41c9d10765b4>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/dada/representative_sequences.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3249d3a-d377-4835-a186-17533dbf1190",
   "metadata": {},
   "source": [
    "## Diversity & Phylogenetics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5be80bf-079a-4aaa-951f-a29a68553acc",
   "metadata": {},
   "source": [
    "[Generate a tree for phylogenetic diversity analyses](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Generate%20a%20tree%20for%20phylogenetic%20diversity%20analyses)\n",
    "\n",
    "From the moving pictures tutorial:\n",
    "> QIIME supports several phylogenetic diversity metrics, including Faith’s Phylogenetic Diversity and weighted and unweighted UniFrac. In addition to counts of features per sample (i.e., the data in the FeatureTable[Frequency] QIIME 2 artifact), these metrics require a rooted phylogenetic tree relating the features to one another. This information will be stored in a Phylogeny[Rooted] QIIME 2 artifact. To generate a phylogenetic tree we will use align-to-tree-mafft-fasttree pipeline from the q2-phylogeny plugin. \r\n",
    "First, the pipeline uses the mafft program to perform a multiple sequence alignment of the sequences in our FeatureData[Sequence] to create a FeatureData[AlignedSequence] QIIME 2 artifact. Next, the pipeline masks (or filters) the alignment to remove positions that are highly variable. These positions are generally considered to add noise to a resulting phylogenetic tree. Following that, the pipeline applies FastTree to generate a phylogenetic tree from the masked alignment. The FastTree program creates an unrooted tree, so in the final step in this section midpoint rooting is applied to place the root of the tree at the midpoint of the longest tip-to-tip distance in the unrooted tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17b6c955-67f3-4f29-b3e9-abbdaf5190dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: ../output/tree/alignment.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: ../output/tree/masked_alignment.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: ../output/tree/tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: ../output/tree/rooted_tree.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences ../output/dada/representative_sequences.qza \\\n",
    "  --output-dir ../output/tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d583891-a192-4fef-a7df-106b4a4a930c",
   "metadata": {},
   "source": [
    "[Diversity](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Alpha%20and%20beta%20diversity%20analysis)\n",
    "\n",
    "QIIME 2’s diversity analyses are available through the `q2-diversity` plugin, which supports computing alpha and beta diversity metrics, applying related statistical tests, and generating interactive visualizations. We’ll first apply the `core-metrics-phylogenetic` method, which rarefies a `FeatureTable Frequency` to a user-specified depth, computes several alpha and beta diversity metrics, and generates principle coordinates analysis (PCoA) plots using Emperor for each of the beta diversity metrics.\n",
    "\n",
    "An important metric to consider when studying microbial ecology is __diversity__. Diversity comes in two flavors: ⍺ (alpha) and β (beta).\r\n",
    "\r\n",
    "Alpha diversity is pretty simple - how diverse is a single sample? You might consider measures like richness and evenness.\r\n",
    "\r\n",
    "![alpha diversity](https://gibbons-lab.github.io/isb_course_2023/16S/assets/alpha_diversity.png)\r\n",
    "\r\n",
    "Beta diversity instead looks at how different two samples are from each other - what taxa are shared, and how their abundances differ.\r\n",
    "\r\n",
    "![beta diversity](https://gibbons-lab.github.io/isb_course_2023/16S/assets/beta_diversity.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e514fb-bd7f-454c-acbc-1aa714df6144",
   "metadata": {},
   "source": [
    "The metrics computed by default are:\n",
    "\r",
    "##### Alpha diversity\r",
    "- \r\n",
    "Shannon’s diversity index (a quantitative measure of community richness- \r\n",
    "\r\n",
    "Observed Features (a qualitative measure of community richne- s)\r\n",
    "\r\n",
    "Faith’s Phylogenetic Dive**rsity (a qualitative measure of community richness that incorporates phylogenetic relationships between the fea**t- res)\r\n",
    "\r\n",
    "Evenness (or Pielou’s Evenness; a measure of community eve##### nness)\r\n",
    "\r\n",
    "Beta - iversity\r\n",
    "\r\n",
    "Jaccard distance (a qualitative measure of community dis- imilarity)\r\n",
    "\r\n",
    "Bray-Curtis distance (a quantitative measure of community d- ssimilarity)\r\n",
    "\r\n",
    "unweighted U**niFrac distance (a qualitative measure of community dissimilarity that incorporates phylogenetic relationships betwe**e-  the features)\r\n",
    "\r\n",
    "weighted** UniFrac distance (a quantitative measure of community dissimilarity that incorporates phylogenetic relationships bet**ween the features) \r"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a79db12-0620-4c43-a365-7e15ae264faa",
   "metadata": {},
   "source": [
    "### Sampling Depth\n",
    "\n",
    "An important parameter that needs to be provided to this code is ` --p-sampling-depth`, which is the even sampling (i.e. rarefaction) depth. Because most diversity metrics are sensitive to different sampling depths across different samples, this script will randomly subsample the counts from each sample to the value provided for this parameter. For am--ple, if you provide `--p-sampling-depth 500`, this step will subsample the counts in each sample without replacement so that each sample in the resulting table has a total count of 500. If the total count for any sample(s) are smaller than this value, those samples will be dropped from the diversity analysis. Choosing this value is tricky. We recommend making your choice by reviewing the information presented in the `table.qzv` file that was created above. Choose a value that is as high as possible (so you retain more sequences per sample) while excluding as few samples as possible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c7695895-ca99-477b-8ad3-d8c5190420ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/c3dd449b-4cce-4766-800a-00f3dd296f0e')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: c3dd449b-4cce-4766-800a-00f3dd296f0e>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## open interactive table visualization\n",
    "Visualization.load('../output/dada2-table.qzv')\n",
    "\n",
    "## navigate to the interactive sample detail tab\n",
    "# move the sampling depth slider as high as you can before excluding any samples \n",
    "# we want the sampling depth to be high, while retaining all 22 samples\n",
    "# this looks like a sampling depth of 10,6727 (09AUG2023, SST) \n",
    "# but maybe we exclude the mock community and bring it up to 163971? (15AUG2023, SST)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dafb35a5-8576-43a4-a82e-e73ff702e5ab",
   "metadata": {},
   "source": [
    "What value would you choose to pass for --p-sampling-depth? \n",
    "- **107,656**\n",
    "How many samples will be excluded from your analysis based on this choice? \n",
    "- **none, all 22 samples are retained**\n",
    "How many total sequences will you be analyzing in the core-metrics-phylogenetic command?\n",
    "- **2,368,410**\n",
    "\n",
    "This represents **40.40%** of the features present across the 22 samples\n",
    "The mock community has the fewest features at **107,656** and is our 'limiting factor' to increase sample depth.\n",
    "Why does the blank have so many features! That is not good... "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f41c70-cc79-4b30-8937-b8e0a6a1fbd9",
   "metadata": {},
   "source": [
    "### Rooted Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efea289d-fe1c-47eb-a549-6179e7655211",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity core-metrics-phylogenetic \\\n",
    "  --i-phylogeny ../output/phylogeny-tree/rooted_tree.qza \\\n",
    "  --i-table ../output/dada2-table.qza \\\n",
    "  --p-sampling-depth 163419 \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --output-dir ../output/diversity-core-denoised"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab4f0ad8-959e-4163-a0ed-34f30972dd7f",
   "metadata": {},
   "source": [
    "### Unweighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f57552fb-d87e-46d8-92f1-2f28bcd493d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/diversity-core/unweighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77451e20-6175-4d81-829e-a2391770b99d",
   "metadata": {},
   "source": [
    "### Weighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eb02ec8-ef94-48cd-84cc-2346a4a71e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/diversity-core/weighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810aeaf1-0d54-4afd-89dd-a5a9ba1c5a39",
   "metadata": {},
   "source": [
    "### Jaccard Emperor PLot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3558a3-558c-4c7e-9548-7c2fc4be2e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/diversity-core/jaccard_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4064ef0f-efe5-48bd-9ebf-8b54817c9edb",
   "metadata": {},
   "source": [
    "### Bray-Curtis Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "235a6b33-14c3-44a6-bacc-99e053c66e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/diversity-core/bray_curtis_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce0cac8e-1e40-4dd5-b25e-375dade9b20e",
   "metadata": {},
   "source": [
    "## Alpha Diversity\n",
    "After computing diversity metrics, we can begin to explore the microbial composition of the samples in the context of the sample metadata. This information is present in the sample metadata file `../rawdata/sample-metadata-verbose.tsv`.\n",
    "\n",
    "We’ll first test for associations between categorical metadata columns and alpha diversity data.\n",
    "We’ll do that here for the Faith Phylogenetic Diversity (a measure of community richness):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "820f2e0b-3a03-4192-b666-f1321e8780da",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/diversity-core/faith_pd_vector.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/faith-pd-group-significance.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacf0ee4-cc0a-4017-9c35-e4409c059127",
   "metadata": {},
   "source": [
    "### Microbial Community Richness: Faith Phylogenetic Diversity\n",
    "\n",
    "##### \n",
    "Which categorical sample metadata columns are most strongly associated with the differences in microbial community richness\n",
    "Temperature Treatment\n",
    "\n",
    "##### Are these differences statistically significat?n\n",
    "\n",
    "Pae\n",
    "- no significantly different groups\n",
    "\n",
    "Temp\n",
    "- ambient (n=12) vs. hot (n=8), pvalue (0.02)\n",
    "  \n",
    "PaeTemp\n",
    "- peak-ambient (n=4) vs. peak-hot (n=4), pvalue (0.04)\n",
    "- env-ambient (n=4) vs. peak-hot (n=4), pvalue (0.04)\n",
    "\n",
    "Colony\n",
    "- no significantly different groups\n",
    "\n",
    "Tank\n",
    "- H1 (n=5) vs KB (n=4), pvalue (0.05)\n",
    "- A2 (n=4) vs H1 (n=5), pvalue (0.05)t?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bc2c581-03af-44d8-909d-a685d5370108",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/faith-pd-group-significance.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5322d40-bde7-471c-9067-0364a9f3cdb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/diversity-core/evenness_vector.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/evenness-group-significance.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7088f177-08e6-4fc8-9a53-010742794998",
   "metadata": {},
   "source": [
    "### Microbial Community Evenness, Alpha diversity , Pielou's Evenness\n",
    "\n",
    "##### Which categorical sample metadata columns are most strongly associated with the differences in microbial community evenness? \n",
    "Nearly everything but Colony.... \n",
    "Surprised by one Tank result: H1 vs H2, I wouldn't have expected those to be different.\n",
    "\n",
    "##### Are these differences statistically significant?\n",
    "Pae\n",
    "- control (n=8) vs peak (n=8), pvalue (0.05)\n",
    "- env (n=4) vs. peak (n=8), pvalue (0.006)\n",
    "\n",
    "Temp\n",
    "- ambient (n=12) vs hot (n=8), pvalue (0.02)\n",
    "\n",
    "PaeTemp\n",
    "- control-ambient (n=4) vs peak-ambient (n=4), pvalue (0.02)\n",
    "- control-ambient (n=4) vs peak-hot (n=4), pvalue (0.02)\n",
    "- env-ambient (n=4) vs peak-ambient (n=4), pvalue (0.02)\n",
    "- env-ambient (n=4) vs peak-hot (n=4), pvalue (0.02)\n",
    "  \n",
    "Colony\n",
    "- no significantly different groups\n",
    "\n",
    "Tank\n",
    "- A1 (n=4) vs H1 (n=5), pvalue (0.02)\n",
    "- A1 (n=4) vs KB (n=4), pvalue (0.04)\n",
    "- A2 (n=4) vs H1 (n=5), pvalue (0.02)\n",
    "- H1 (n=5) vs H2 (n=3), pvalue (0.05)\n",
    "- H1 (n=5) vs KB (n=4), pvalue (0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ed65296-ac10-440b-8819-89d409acc37d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/evenness-group-significance.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123d925c-3c85-406f-834c-0ef6cf32d3b9",
   "metadata": {},
   "source": [
    "## Beta Diversity\n",
    "\n",
    "Next we’ll analyze sample composition in the context of categorical metadata using PERMANOVA (first described in Anderson (2001)) using the beta-group-significance command. The following commands will test whether distances between samples within a group, such as samples from the same body site (e.g., gut), are more similar to each other then they are to samples from the other groups (e.g., tongue, left palm, and right palm). If you call this command with the --p-pairwise parameter, as we’ll do here, it will also perform pairwise tests that will allow you to determine which specific pairs of groups (e.g., tongue and gut) differ from one another, if any. This command can be slow to run, especially when passing --p-pairwise, since it is based on permutation tests. So, unlike the previous commands, we’ll run beta-group-significance on specific columns of metadata that we’re interested in exploring, rather than all metadata columns to which it is applicable. Here we’ll apply this to our unweighted UniFrac distances, using two sample metadata columns, as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a10501-6323-4cba-be60-be6e370a7077",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-core/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-slim.tsv \\\n",
    "  --m-metadata-column Pae \\\n",
    "  --o-visualization ../output/diversity-core/unweighted-unifrac-body-site-significance.qzv \\\n",
    "  --p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-core/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-slim.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/diversity-core/unweighted-unifrac-body-site-significance.qzv \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa1712a0-d269-4da9-86ba-80432e67a8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-core/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/diversity-core/unweighted-unifrac-body-site-significance.qzv \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a6a5115-44b4-4c07-bc8f-6088c11926c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-core/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/diversity-core/unweighted-unifrac-subject-group-significance.qzv \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e9a30b-96cf-4455-b040-2c21e93fd171",
   "metadata": {},
   "source": [
    "## [Atacama Soil Microbiome: Questions to Guide Data Analysis](https://docs.qiime2.org/2023.5/tutorials/atacama-soils/#paired-end-read-analysis-commands:~:text=Questions%20to%20guide%20data%20analysis)\n",
    "What sample metadata or combinations of sample metadata are most strongly associated with the differences in microbial composition of the samples? Are these associations stronger with unweighted UniFrac or with Bray-Curtis? Based on what you know about these metrics, what does that difference suggest? For exploring associations between continuous metadata and sample composition, the commands qiime metadata distance-matrix in combination with qiime diversity mantel and qiime diversity bioenv will be useful. These were not covered in the Moving Pictures tutorial, but you can learn about them by running them with the `--help` parameter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8380a16-f76f-4232-8f6c-aa99e14abf8b",
   "metadata": {},
   "source": [
    "## Alpha rarefaction plotting\n",
    "In this section we’ll explore alpha diversity as a function of sampling depth using the qiime diversity alpha-rarefaction visualizer. This visualizer computes one or more alpha diversity metrics at multiple sampling depths, in steps between 1 (optionally controlled with --p-min-depth) and the value provided as --p-max-depth. At each sampling depth step, 10 rarefied tables will be generated, and the diversity metrics will be computed for all samples in the tables. The number of iterations (rarefied tables computed at each sampling depth) can be controlled with --p-iterations. Average diversity values will be plotted for each sample at each even sampling depth, and samples can be grouped based on metadata in the resulting visualization if sample metadata is provided with the --m-metadata-file parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfcb4d2b-a044-4434-9980-37282f97c045",
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime diversity alpha-rarefaction \\\n",
    "  --i-table ../output/dada2-table.qza \\\n",
    "  --i-phylogeny ../output/phylogeny-tree/rooted_tree.qza \\\n",
    "  --p-max-depth 107656 \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/alpha-rarefaction.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2a1e0bd-d356-4c77-a56e-3363864f7d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "Visualization.load('../output/alpha-rarefaction.qzv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
