{
 "cells": [
  {
   "cell_type": "raw",
   "id": "e15ca787-fd40-4888-94e7-b78a721b05b9",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Microbiome Sequence Analysis\"\n",
    "subtitle: \"Importing 16S Sequence Data\"\n",
    "author: \"Sarah Tanja\"\n",
    "format: html\n",
    "editor: visual\n",
    "toc: true\n",
    "toc-title: Contents <i class=\"bi bi-bookmark-heart\"></i>\n",
    "toc-depth: 5\n",
    "toc-location: left\n",
    "reference-location: margin\n",
    "citation-location: margin\n",
    "bibliography: ./references/qiime2.bib\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "122fda22-a268-433a-b619-ec18f7946413",
   "metadata": {},
   "source": [
    "## Overview\n",
    "### Method\n",
    "\n",
    "### Data download\n",
    "I downloaded two .zip folders from a Dropbox link; one had raw data files and the other had analysis pipeline files generated by [Mr. DNA Lab](https://www.mrdnalab.com/) Molecular Research. I unzipped the folders and uploaded the Mr. DNA analysis pipeline files into the `coral-pae-temp/analysis/microbiome/mrdna` directory, and the `sample-metadata.tsv` and `demux` folder into `coral-pae-temp/analysis/microbiome/rawdata` directory.  \n",
    "\n",
    "### Data type\n",
    "\n",
    "Here I am working with the raw data files located in `coral-pae-temp/analysis/microbiome/data/demux`. In the `demux` folder are two `fastq.gz` files for each of the 22 samples, one for the forward read and one for the reverse read. \n",
    "\n",
    "The `fastq.gz` file name includes the sample identifier and should look like `4.Ea_S1_L001_R1_001.fastq.gz`. \n",
    "The underscore-separated fields in this file name are:\n",
    "\n",
    "1.  the sample identifier,\n",
    "\n",
    "2.  the barcode sequence or a barcode identifier,\n",
    "\n",
    "3.  the lane number,\n",
    "\n",
    "4.  the direction of the read (i.e. R1 or R2, because these are paired-end reads), and\n",
    "\n",
    "5.  the set number.\n",
    "   \n",
    "\n",
    "The `fastq.gz` files are **Demultiplexed** (aka **Demuxed**) sequences that still have the forward and reverse primers in the sequences.\n",
    "\n",
    "-   The Raw Data is **demultiplexed**\n",
    "\n",
    "-   A R1 and R2 fastq.gz file has been generated for each individual sample\n",
    "\n",
    "-   All forward reads are binned into the R1 fastq.gz files\n",
    "\n",
    "-   All reverse reads are binned into the R2 fastq.gz files\n",
    "\n",
    "-   Other than demultiplexing; you can consider the Raw Data on BaseSpace as untouched (**The Forward and Reverse Primer Sequences have not been removed**)\n",
    "\n",
    "### Data process\n",
    "Here I follow the QIIME2 [Casava 1.8 paired-end demultiplexed fastq](https://docs.qiime2.org/2023.5/tutorials/importing/#:~:text=Casava%201.8%20paired%2Dend%20demultiplexed%20fastq) tutorial example on importing data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7997d9-10ca-4323-85e2-676ee96f60b5",
   "metadata": {},
   "source": [
    "## Python 3 API import qiime plugins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a593cfd4-fcb7-4a42-8440-457f99e65015",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18dc8856-97ef-42d6-9f4c-3cdd23ea0f83",
   "metadata": {},
   "source": [
    "## Metadata\n",
    "Make a table of the metadata.\n",
    "Here I added columns 'Pae', 'Temp', 'PeaTemp', 'Colony', and 'Tank' to the original `sample-metadata.tsv` file provided to me by Mr. DNA and renamed it `sample-metadata-verbose.tsv`\n",
    "This was a bit of a process... I had to:\n",
    "upload the `sample-metadata.tsv` to Excel\n",
    "edit the metadata by adding the above columns and values\n",
    "save it as a csv\n",
    "open it in a text editor\n",
    "search for all ',' commas, and find&replace them with 'TAB' symbols\n",
    "save as a tab separated file `.tsv`\n",
    "upload it back into the `coral-pae-temp/analysis/microbiome/rawdata` folder\n",
    "\n",
    "At first I had named the new columns 'pae', 'temp', etc. with lower case... for some reason this was a problem later on and the interactive emperor plots wouldn't recognize the new columns. When I changed the column names to CamelCase to match the others, it worked. The `qiime2` docs indicate that metadata formatted with an Identifier Column such as `#Sample ID` is [case-sensitive](https://docs.qiime2.org/2023.5/tutorials/metadata/#metadata-formatting-requirements:~:text=feature%2Did-,Case%2Dsensitive,-(these%20are%20mostly)\n",
    "\n",
    "Later I learned you can edit .tsv files directly in Jupyter Notebooks using pandas\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c420fcd4-0a0c-4fa7-a36f-4a1d4069b4ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/sample-metadata-verbose.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file ../data/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/sample-metadata-verbose.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fd415645-5a55-4d00-8dfd-e2d2246a3c3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/8959decd-55e6-4113-94a5-1cea20c31692')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 8959decd-55e6-4113-94a5-1cea20c31692>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/sample-metadata-verbose.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9afbd41",
   "metadata": {},
   "source": [
    "Import Sequences into qiime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dbb298fe-3e96-4f98-9060-408d3b24511a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported ../data/demux as CasavaOneEightSingleLanePerSampleDirFmt to ../output/demux-paired-end.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path ../data/demux \\\n",
    "  --input-format CasavaOneEightSingleLanePerSampleDirFmt \\\n",
    "  --output-path ../output/demux-paired-end.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17971af4-e264-4f18-b33a-3e774fdf7295",
   "metadata": {},
   "source": [
    "The `demux-paired-end.qza` artifact contains raw, demultiplexed sequences that still have forward and reverse primers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cc3897b-7598-406d-8813-da555c48ebb8",
   "metadata": {},
   "source": [
    "## Trim primers from paired-end sequences using `cutadapt`\n",
    "\n",
    "> \"The PCR primers (F515/R806) were developed against the V4 region of the 16S rRNA, which we determined would yield optimal community clustering with reads of this length using a procedure similar to that of ref. 15. [For reference, this primer pair amplifies the region 533–786 in the Escherichia coli strain 83972 sequence (greengenes accession no. prokMSA_id:470367).] The reverse PCR primer is barcoded with a 12-base error correcting Golay code to facilitate multiplexing of up to ≈1,500 samples per lane, and both PCR primers contain sequencer adapter regions.\" - (Caporasco et al. 2011)\n",
    "\n",
    "Caporaso, J. G., Lauber, C. L., Walters, W. A., Berg-Lyons, D., Lozupone, C. A., Turnbaugh, P. J., Fierer, N., & Knight, R. (2011). Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample. Proceedings of the National Academy of Sciences, 108(supplement_1), 4516–4522. https://doi.org/10.1073/pnas.1000080107\n",
    "\n",
    "> \"The V4 variable region of the 16S rRNA gene was amplified using the 515F (5′-­GTGCCAGCMGCCGCGGTAA-­3′) and 806R (5′-­GGACTACHVGGGTWTCTAAT-­3′) primer set (Caporaso et al. 2011). ” - (Brown et al. 2021)\n",
    "\n",
    "Brown, Tanya, Dylan Sonett, Jesse R. Zaneveld, and Jacqueline L. Padilla-Gamiño. 2021. “Characterization of the Microbiome and Immune Response in Corals with Chronic Montipora White Syndrome.” Molecular Ecology 30 (11): 2591–2606. https://doi.org/10.1111/mec.15899."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "967c62c9-a86a-48f6-9d5f-6d6a5501e1ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved SampleData[PairedEndSequencesWithQuality] to: ../output/demux-trimmed.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime cutadapt trim-paired \\\n",
    "  --i-demultiplexed-sequences ../output/demux-paired-end.qza \\\n",
    "  --p-cores 4 \\\n",
    "  --p-front-f GTGYCAGCMGCCGCGGTAA \\\n",
    "  --p-front-r GGACTACNVGGGTWTCTAAT \\\n",
    "  --o-trimmed-sequences ../output/demux-trimmed.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00901ad",
   "metadata": {},
   "source": [
    "## Visualize trimmed & demultiplexed sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "69b6a077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/demux-trimmed-summary.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime demux summarize \\\n",
    "  --i-data ../output/demux-trimmed.qza \\\n",
    "  --o-visualization ../output/demux-trimmed-summary.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "da27f217-31b2-4cb5-8edb-6fd6bd70f48f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/e172f028-ef64-4c83-befe-d10572d906c8')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: e172f028-ef64-4c83-befe-d10572d906c8>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/demux-trimmed-summary.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873d9098-a910-4845-ade0-8b55c7622239",
   "metadata": {},
   "source": [
    "## [Denoise with DADA2](https://docs.qiime2.org/2023.5/tutorials/moving-pictures/#sequence-quality-control-and-feature-table-construction:~:text=with%20QIIME%201.-,Option%201%3A%20DADA2%C2%B6,-DADA2%20is%20a)\n",
    "\n",
    "[DADA2](https://pubmed.ncbi.nlm.nih.gov/27214047/) is a pipeline for detecting and correcting (where possible) Illumina amplicon sequence data. \n",
    "\n",
    "As implemented in the q2-dada2 plugin, this quality control process will additionally filter any phiX reads (commonly present in marker gene Illumina sequence data) that are identified in the sequencing data, and will filter chimeric sequences.  \n",
    "The dada2 denoise-paired method requires four parameters that are used in quality filtering:  \n",
    "\n",
    "    `--p-trim-left-f m`, which trims off the first m bases of each sequence in the forward reads\n",
    "    `--p-trim-left-r n`, which trims off the first m bases of each sequence in the reverse reads\n",
    "    `--p-trunc-len-f o`, which truncates each sequence at position o in the forward reads\n",
    "    `--p-trunc-len-f p`, which truncates each sequence at position p in the reverse reads  \n",
    "    \n",
    "This allows the user to remove low quality regions of the sequences.  \n",
    "\n",
    "What is a 'good' quality score?  \n",
    "\n",
    "In QIIME 2's interactive quality plots, the quality scores typically range from 0 to 40. Quality scores reflect the accuracy of base calls in sequencing data, with higher scores indicating higher accuracy. The most common quality score scale used in modern sequencing technologies is the Phred scale.\r",
    "<br>\n",
    "\r\n",
    "In the Phred sceal-<br> \r\n",
    "\r\n",
    "A quality score of 10 corresponds to a 1 in 10 chance of an incorrect base call (90% aate -<br> y).\r\n",
    "A quality score of 20 corresponds to a 1 in 100 chance of an incorrect base call (99%rate -<br> cy).\r\n",
    "A quality score of 30 corresponds to a 1 in 1000 chance of an incorrect base call  (9curate..<br>%<br>\n",
    "accur\n",
    "A \"good\" quality score in this context depends on your specific analysis goals and the sequencing platform you're u\n",
    "sing. However, many researchers consider quality scores above 20 to be generally acceptable for downstreaanal**ysis. Scores above 30 are often seen as very ha qu*. In theew the Interactive Quality Plot tab in the `demux-trimmed-summary.qzv` file that was generated by `qiime demux summ lots, we see that the quality scores of the bases are high, between a score of 11 in the lowest 2nd percentile and a score of 37 in the bottom 25th percentile and higher. So we won’t trim any bases from the beginning of thences. \n",
    "<br>er.  \n",
    "\n",
    "[Denoising tips from Greg Caporasco](https://docs.qiime2.org/jupyterbooks/cancer-microbiome-intervention-tutorial/020-tutorial-upstream/040-denoising.html) using `qiime dada2 denoise-pa<br>iForward: 230\n",
    "<br>\n",
    "Reverse: 2- R<br>, e31\n",
    "   \n",
    "*This next command may take up to 10min to run and is the slowest step*t step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc5e87b8-0b3b-46ab-b1e2-4291fb1ed11b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada.R --input_directory /tmp/tmp7xn9n2ih/forward --input_directory_reverse /tmp/tmp7xn9n2ih/reverse --output_path /tmp/tmp7xn9n2ih/output.tsv.biom --output_track /tmp/tmp7xn9n2ih/track.tsv --filtered_directory /tmp/tmp7xn9n2ih/filt_f --filtered_directory_reverse /tmp/tmp7xn9n2ih/filt_r --truncation_length 230 --truncation_length_reverse 231 --trim_left 0 --trim_left_reverse 0 --max_expected_errors 2.0 --max_expected_errors_reverse 2.0 --truncation_quality_score 2 --min_overlap 12 --pooling_method independent --chimera_method consensus --min_parental_fold 1.0 --allow_one_off False --num_threads 20 --learn_min_reads 1000000\n",
      "\n",
      "R version 4.2.3 (2023-03-15) \n",
      "Loading required package: Rcpp\n",
      "\u001b[?25hDADA2: 1.26.0 / Rcpp: 1.0.10 / RcppParallel: 5.1.6 \n",
      "\u001b[?25h\u001b[?25h2) Filtering \u001b[?25h\u001b[?25h......................\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "\u001b[?25h\u001b[?25h3) Learning Error Rates\n",
      "\u001b[?25h240503870 total bases in 1045669 reads from 3 samples will be used for learning the error rates.\n"
     ]
    }
   ],
   "source": [
    "!qiime dada2 denoise-paired \\\n",
    "  --i-demultiplexed-seqs ../output/demux-trimmed.qza \\\n",
    "  --p-trim-left-f 0 \\\n",
    "  --p-trim-left-r 0 \\\n",
    "  --p-trunc-len-f 230 \\\n",
    "  --p-trunc-len-r 231 \\\n",
    "  --p-n-threads 20 \\\n",
    "  --output-dir ../output/dada2 --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3ff89ca9-6ec3-4d24-962f-c3f21f8249d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/denoising-stats-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime metadata tabulate \\\n",
    "  --m-input-file ../output/blank-mock/denoising-stats-bm.qza \\\n",
    "  --o-visualization ../output/blank-mock/denoising-stats-bm.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "043972a4-d961-43d4-b008-622b0b5a94ab",
   "metadata": {},
   "source": [
    "### Visualize Denoising Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "62530108-5547-4a61-941e-ad81864dc0ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/930df4f9-2171-4ac4-b487-8cf21e5baa4c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 930df4f9-2171-4ac4-b487-8cf21e5baa4c>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/denoising-stats-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "586afdc0-6bde-4b61-a10e-886d1de3a839",
   "metadata": {},
   "source": [
    "## Summarize & tabulate the feature table\n",
    "After the quality filtering step completes, you’ll want to explore the resulting data. You can do this using the following two commands, which will create visual summaries of the data. The `feature-table summarize` command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics. The `feature-table tabulate-seqs` command will provide a mapping of feature IDs to sequences, and provide links to easily BLAST each sequence against the NCBI nt database.\n",
    "\n",
    "### feature-table summarize\n",
    "The feature-table summarize command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics. Feature tables in QIIME 2 represent the abundance of different biological features (such as bacterial taxa or OTUs) across samples.\n",
    " In this command:\n",
    "\n",
    "--i-table tab.qzal`es the input feature table in QIIME 2 artifact format (.qza file) that you want to sumri\r\n",
    "\r\n",
    "--o-visualizattable.qzv tfies the output visualization in QIIME 2 artifact format (.qzv file) that will con the summaes.\r\n",
    "\r\n",
    "--m-sample-metadata-file sampldata.tsve-cifies the metadata file (usually in tab-separated values format) that contains additional information about the samples in your featble  and m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "18f2f01e-f8cc-4feb-b0e6-d4316ef9a216",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/table-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table summarize \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --o-visualization ../output/blank-mock/table-bm.qzv \\\n",
    "  --m-sample-metadata-file ../rawdata/sample-metadata-verbose.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f74905c7-7813-44bd-a27d-0350d9866e36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/ee2b6637-f808-40bd-8df0-b1352d5a6679')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: ee2b6637-f808-40bd-8df0-b1352d5a6679>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/table-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7144554a-99b0-4fbd-8a79-606433ad2fb7",
   "metadata": {},
   "source": [
    "### feature-table tabulate-seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "16801dde-0810-47ae-ba30-76867ef1be15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/rep-seqs-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table tabulate-seqs \\\n",
    "  --i-data ../output/blank-mock/rep-seqs-bm.qza \\\n",
    "  --o-visualization ../output/blank-mock/rep-seqs-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0e800ea2-04c3-4594-a96c-8689bf39b282",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/de72ef09-38ca-4d56-ae25-83425c5ed3d3')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: de72ef09-38ca-4d56-ae25-83425c5ed3d3>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/rep-seqs-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5be80bf-079a-4aaa-951f-a29a68553acc",
   "metadata": {},
   "source": [
    "## [Generate a tree for phylogenetic diversity analyses](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Generate%20a%20tree%20for%20phylogenetic%20diversity%20analyses)\n",
    "\n",
    "From the moving pictures tutorial:\n",
    "> QIIME supports several phylogenetic diversity metrics, including Faith’s Phylogenetic Diversity and weighted and unweighted UniFrac. In addition to counts of features per sample (i.e., the data in the FeatureTable[Frequency] QIIME 2 artifact), these metrics require a rooted phylogenetic tree relating the features to one another. This information will be stored in a Phylogeny[Rooted] QIIME 2 artifact. To generate a phylogenetic tree we will use align-to-tree-mafft-fasttree pipeline from the q2-phylogeny plugin. \r\n",
    "First, the pipeline uses the mafft program to perform a multiple sequence alignment of the sequences in our FeatureData[Sequence] to create a FeatureData[AlignedSequence] QIIME 2 artifact. Next, the pipeline masks (or filters) the alignment to remove positions that are highly variable. These positions are generally considered to add noise to a resulting phylogenetic tree. Following that, the pipeline applies FastTree to generate a phylogenetic tree from the masked alignment. The FastTree program creates an unrooted tree, so in the final step in this section midpoint rooting is applied to place the root of the tree at the midpoint of the longest tip-to-tip distance in the unrooted tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "17b6c955-67f3-4f29-b3e9-abbdaf5190dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[AlignedSequence] to: ../output/blank-mock/phylogeny-tree-bm/alignment.qza\u001b[0m\n",
      "\u001b[32mSaved FeatureData[AlignedSequence] to: ../output/blank-mock/phylogeny-tree-bm/masked_alignment.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Unrooted] to: ../output/blank-mock/phylogeny-tree-bm/tree.qza\u001b[0m\n",
      "\u001b[32mSaved Phylogeny[Rooted] to: ../output/blank-mock/phylogeny-tree-bm/rooted_tree.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences ../output/blank-mock/rep-seqs-bm.qza \\\n",
    "  --output-dir ../output/blank-mock/phylogeny-tree-bm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d583891-a192-4fef-a7df-106b4a4a930c",
   "metadata": {},
   "source": [
    "## [Alpha & Beta diversity](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Alpha%20and%20beta%20diversity%20analysis)\n",
    "\n",
    "![alpha-vs-beta](https://commercio.nyc3.digitaloceanspaces.com/goldbio-2018/pages/2.%20Alpha%20vs%20Beta%20diversity.png)\n",
    "\n",
    "QIIME 2’s diversity analyses are available through the `q2-diversity` plugin, which supports computing alpha and beta diversity metrics, applying related statistical tests, and generating interactive visualizations. We’ll first apply the `core-metrics-phylogenetic` method, which rarefies a `FeatureTable Frequency` to a user-specified depth, computes several alpha and beta diversity metrics, and generates principle coordinates analysis (PCoA) plots using Emperor for each of the beta diversity metrics. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e514fb-bd7f-454c-acbc-1aa714df6144",
   "metadata": {},
   "source": [
    "The metrics computed by default are:\n",
    "\r",
    "##### Alpha diversity\r",
    "- \r\n",
    "Shannon’s diversity index (a quantitative measure of community richness- \r\n",
    "\r\n",
    "Observed Features (a qualitative measure of community richne- s)\r\n",
    "\r\n",
    "Faith’s Phylogenetic Dive**rsity (a qualitative measure of community richness that incorporates phylogenetic relationships between the fea**t- res)\r\n",
    "\r\n",
    "Evenness (or Pielou’s Evenness; a measure of community eve##### nness)\r\n",
    "\r\n",
    "Beta - iversity\r\n",
    "\r\n",
    "Jaccard distance (a qualitative measure of community dis- imilarity)\r\n",
    "\r\n",
    "Bray-Curtis distance (a quantitative measure of community d- ssimilarity)\r\n",
    "\r\n",
    "unweighted U**niFrac distance (a qualitative measure of community dissimilarity that incorporates phylogenetic relationships betwe**e-  the features)\r\n",
    "\r\n",
    "weighted** UniFrac distance (a quantitative measure of community dissimilarity that incorporates phylogenetic relationships bet**ween the features) \r"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4951088-76dc-4028-8513-a5521d85b35d",
   "metadata": {},
   "source": [
    "\r\n",
    "An important parameter that needs to be prii`ded to this scri ` i`s --p-sampling-dep`th, which is the even sampling (i.e. rarefaction) depth. Because most diversity metrics are sensitive to different sampling depths across different samples, this script will randomly subsample the counts from each sample to the value provided for this parameter. For am--ple, if you provide `--p-sampling-depth 5`00, this step will subsample the counts in each sample without replacement so that each sample in the resulting table has a total count of 500. If the total count for any sample(s) are smaller than this value, those samples will be dropped from the diversity analysis. Choosing this value is tricky. We recommend making your choice by reviewing the infora`tion prs`ented in th`e table.q`zv file that was created above. Choose a value that is as high as possible (so you retain more sequences per sampl)while excluding as few w samples as possiblefe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a79db12-0620-4c43-a365-7e15ae264faa",
   "metadata": {},
   "source": [
    "### Sampling Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c7695895-ca99-477b-8ad3-d8c5190420ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/ee2b6637-f808-40bd-8df0-b1352d5a6679')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: ee2b6637-f808-40bd-8df0-b1352d5a6679>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## open interactive table visualization\n",
    "Visualization.load('../output/blank-mock/table-bm.qzv')\n",
    "\n",
    "## navigate to the interactive sample detail tab\n",
    "# move the sampling depth slider as high as you can before excluding any samples \n",
    "# we want the sampling depth to be high, while retaining all 22 samples\n",
    "# this looks like a sampling depth of 10,6727 (09AUG2023, SST) \n",
    "# but maybe we exclude the mock community and bring it up to 163971? (15AUG2023, SST)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dafb35a5-8576-43a4-a82e-e73ff702e5ab",
   "metadata": {},
   "source": [
    "What value would you choose to pass for --p-sampling-depth? \n",
    "- **104,183**\n",
    "How many samples will be excluded from your analysis based on this choice? \n",
    "- **none, all 22 samples are retained**\n",
    "How many total reads will you be analyzing in the core-metrics-phylogenetic command?\n",
    "- **2,292,026**\n",
    "\n",
    "This represents **42.33%** of the features present across the 22 samples\n",
    "The mock community has the fewest reads at **104,183** and is our 'limiting factor' to increase sample depth.\n",
    "Why does the blank have so many features! That is not good... "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f41c70-cc79-4b30-8937-b8e0a6a1fbd9",
   "metadata": {},
   "source": [
    "### Rooted Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "efea289d-fe1c-47eb-a549-6179e7655211",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: ../output/blank-mock/diversity-core-bm/rarefied_table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/blank-mock/diversity-core-bm/faith_pd_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/blank-mock/diversity-core-bm/observed_features_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/blank-mock/diversity-core-bm/shannon_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/blank-mock/diversity-core-bm/evenness_vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/blank-mock/diversity-core-bm/unweighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/blank-mock/diversity-core-bm/weighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/blank-mock/diversity-core-bm/jaccard_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/blank-mock/diversity-core-bm/bray_curtis_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/blank-mock/diversity-core-bm/unweighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/blank-mock/diversity-core-bm/weighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/blank-mock/diversity-core-bm/jaccard_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/blank-mock/diversity-core-bm/bray_curtis_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/diversity-core-bm/unweighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/diversity-core-bm/weighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/diversity-core-bm/jaccard_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/diversity-core-bm/bray_curtis_emperor.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity core-metrics-phylogenetic \\\n",
    "  --i-phylogeny ../output/blank-mock/phylogeny-tree-bm/rooted_tree.qza \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --p-sampling-depth 104183 \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --output-dir ../output/blank-mock/diversity-core-bm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab4f0ad8-959e-4163-a0ed-34f30972dd7f",
   "metadata": {},
   "source": [
    "### Unweighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f57552fb-d87e-46d8-92f1-2f28bcd493d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/fda23875-f6d6-4be3-ba49-51cc9ccd0e3d')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: fda23875-f6d6-4be3-ba49-51cc9ccd0e3d>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/diversity-core-bm/unweighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77451e20-6175-4d81-829e-a2391770b99d",
   "metadata": {},
   "source": [
    "### Weighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1eb02ec8-ef94-48cd-84cc-2346a4a71e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/03d37f8b-cf84-40c7-8693-8140b1a2efa8')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 03d37f8b-cf84-40c7-8693-8140b1a2efa8>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/diversity-core-bm/weighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810aeaf1-0d54-4afd-89dd-a5a9ba1c5a39",
   "metadata": {},
   "source": [
    "### Jaccard Emperor PLot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ff3558a3-558c-4c7e-9548-7c2fc4be2e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/fc5d2682-c523-4547-8183-b4c7119e78da')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: fc5d2682-c523-4547-8183-b4c7119e78da>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/diversity-core-bm/jaccard_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4064ef0f-efe5-48bd-9ebf-8b54817c9edb",
   "metadata": {},
   "source": [
    "### Bray-Curtis Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "235a6b33-14c3-44a6-bacc-99e053c66e89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/6877eff0-1940-4ef5-83f0-8210c95a624a')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 6877eff0-1940-4ef5-83f0-8210c95a624a>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/diversity-core-bm/bray_curtis_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce0cac8e-1e40-4dd5-b25e-375dade9b20e",
   "metadata": {},
   "source": [
    "## Alpha Diversity\n",
    "After computing diversity metrics, we can begin to explore the microbial composition of the samples in the context of the sample metadata. This information is present in the sample metadata file `../rawdata/sample-metadata-verbose.tsv`.\n",
    "\n",
    "We’ll first test for associations between categorical metadata columns and alpha diversity data.\n",
    "We’ll do that here for the Faith Phylogenetic Diversity (a measure of community richness):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "820f2e0b-3a03-4192-b666-f1321e8780da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/faith-pd-group-significance-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/blank-mock/diversity-core-bm/faith_pd_vector.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/blank-mock/faith-pd-group-significance-bm.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacf0ee4-cc0a-4017-9c35-e4409c059127",
   "metadata": {},
   "source": [
    "### Microbial Community Richness: Faith Phylogenetic Diversity\n",
    "\n",
    "##### \n",
    "Which categorical sample metadata columns are most strongly associated with the differences in microbial community richness\n",
    "Temperature Treatment\n",
    "\n",
    "##### Are these differences statistically significat?n\n",
    "\n",
    "Pae\n",
    "- no significantly different groups\n",
    "\n",
    "Temp\n",
    "- ambient (n=12) vs. hot (n=8), pvalue (0.02)\n",
    "  \n",
    "PaeTemp\n",
    "- peak-ambient (n=4) vs. peak-hot (n=4), pvalue (0.04)\n",
    "- env-ambient (n=4) vs. peak-hot (n=4), pvalue (0.04)\n",
    "\n",
    "Colony\n",
    "- no significantly different groups\n",
    "\n",
    "Tank\n",
    "- H1 (n=5) vs KB (n=4), pvalue (0.05)\n",
    "- A2 (n=4) vs H1 (n=5), pvalue (0.05)t?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7bc2c581-03af-44d8-909d-a685d5370108",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/3f1b74b4-3d27-4399-abb6-10a947f7631d')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 3f1b74b4-3d27-4399-abb6-10a947f7631d>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/faith-pd-group-significance-bm.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e5322d40-bde7-471c-9067-0364a9f3cdb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/evenness-group-significance-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/blank-mock/diversity-core-bm/evenness_vector.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/blank-mock/evenness-group-significance-bm.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7088f177-08e6-4fc8-9a53-010742794998",
   "metadata": {},
   "source": [
    "### Microbial Community Evenness, Alpha diversity , Pielou's Evenness\n",
    "\n",
    "##### Which categorical sample metadata columns are most strongly associated with the differences in microbial community evenness? \n",
    "Nearly everything but Colony.... \n",
    "Surprised by one Tank result: H1 vs H2, I wouldn't have expected those to be different.\n",
    "\n",
    "##### Are these differences statistically significant?\n",
    "Pae\n",
    "- control (n=8) vs peak (n=8), pvalue (0.05)\n",
    "- env (n=4) vs. peak (n=8), pvalue (0.006)\n",
    "\n",
    "Temp\n",
    "- ambient (n=12) vs hot (n=8), pvalue (0.02)\n",
    "\n",
    "PaeTemp\n",
    "- control-ambient (n=4) vs peak-ambient (n=4), pvalue (0.02)\n",
    "- control-ambient (n=4) vs peak-hot (n=4), pvalue (0.02)\n",
    "- env-ambient (n=4) vs peak-ambient (n=4), pvalue (0.02)\n",
    "- env-ambient (n=4) vs peak-hot (n=4), pvalue (0.02)\n",
    "  \n",
    "Colony\n",
    "- no significantly different groups\n",
    "\n",
    "Tank\n",
    "- A1 (n=4) vs H1 (n=5), pvalue (0.02)\n",
    "- A1 (n=4) vs KB (n=4), pvalue (0.04)\n",
    "- A2 (n=4) vs H1 (n=5), pvalue (0.02)\n",
    "- H1 (n=5) vs H2 (n=3), pvalue (0.05)\n",
    "- H1 (n=5) vs KB (n=4), pvalue (0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4ed65296-ac10-440b-8819-89d409acc37d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/9be2acf0-4c8d-418e-9248-73a750bbf39a')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 9be2acf0-4c8d-418e-9248-73a750bbf39a>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/evenness-group-significance-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123d925c-3c85-406f-834c-0ef6cf32d3b9",
   "metadata": {},
   "source": [
    "## Beta Diversity (There is a problem with the metadata I can't figure out here)\n",
    "\n",
    "Next we’ll analyze sample composition in the context of categorical metadata using PERMANOVA (first described in Anderson (2001)) using the beta-group-significance command. The following commands will test whether distances between samples within a group, such as samples from the same body site (e.g., gut), are more similar to each other then they are to samples from the other groups (e.g., tongue, left palm, and right palm). If you call this command with the --p-pairwise parameter, as we’ll do here, it will also perform pairwise tests that will allow you to determine which specific pairs of groups (e.g., tongue and gut) differ from one another, if any. This command can be slow to run, especially when passing --p-pairwise, since it is based on permutation tests. So, unlike the previous commands, we’ll run beta-group-significance on specific columns of metadata that we’re interested in exploring, rather than all metadata columns to which it is applicable. Here we’ll apply this to our unweighted UniFrac distances, using two sample metadata columns, as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b7a10501-6323-4cba-be60-be6e370a7077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime diversity beta-group-significance\u001b[0m [OPTIONS]\n",
      "\n",
      "  Determine whether groups of samples are significantly different from one\n",
      "  another using a permutation-based statistical test.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m     Matrix of distances between pairs of samples.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--m-metadata-file\u001b[0m METADATA\n",
      "  \u001b[94m\u001b[4m--m-metadata-column\u001b[0m COLUMN  \u001b[32mMetadataColumn[Categorical]\u001b[0m\n",
      "                       Categorical sample metadata column.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--p-method\u001b[0m TEXT \u001b[32mChoices('permanova', 'anosim', 'permdisp')\u001b[0m\n",
      "                       The group significance test to be applied.\n",
      "                                                        \u001b[35m[default: 'permanova']\u001b[0m\n",
      "  \u001b[94m--p-pairwise\u001b[0m / \u001b[94m--p-no-pairwise\u001b[0m\n",
      "                       Perform pairwise tests between all pairs of groups in\n",
      "                       addition to the test across all groups. This can be\n",
      "                       very slow if there are a lot of groups in the metadata\n",
      "                       column.                                \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-permutations\u001b[0m INTEGER\n",
      "                       The number of permutations to be run when computing\n",
      "                       p-values.                                \u001b[35m[default: 999]\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-visualization\u001b[0m VISUALIZATION\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH    Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m  Display verbose output to stdout and/or stderr during\n",
      "                       execution of this action. Or silence output if\n",
      "                       execution is successful (silence is golden).\n",
      "  \u001b[94m--example-data\u001b[0m PATH  Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m          Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m               Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Missing option '--m-metadata-file'.\u001b[0m\n",
      "\u001b[0mUsage: \u001b[94mqiime diversity beta-group-significance\u001b[0m [OPTIONS]\n",
      "\n",
      "  Determine whether groups of samples are significantly different from one\n",
      "  another using a permutation-based statistical test.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-distance-matrix\u001b[0m ARTIFACT\n",
      "    \u001b[32mDistanceMatrix\u001b[0m     Matrix of distances between pairs of samples.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--m-metadata-file\u001b[0m METADATA\n",
      "  \u001b[94m\u001b[4m--m-metadata-column\u001b[0m COLUMN  \u001b[32mMetadataColumn[Categorical]\u001b[0m\n",
      "                       Categorical sample metadata column.          \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--p-method\u001b[0m TEXT \u001b[32mChoices('permanova', 'anosim', 'permdisp')\u001b[0m\n",
      "                       The group significance test to be applied.\n",
      "                                                        \u001b[35m[default: 'permanova']\u001b[0m\n",
      "  \u001b[94m--p-pairwise\u001b[0m / \u001b[94m--p-no-pairwise\u001b[0m\n",
      "                       Perform pairwise tests between all pairs of groups in\n",
      "                       addition to the test across all groups. This can be\n",
      "                       very slow if there are a lot of groups in the metadata\n",
      "                       column.                                \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-permutations\u001b[0m INTEGER\n",
      "                       The number of permutations to be run when computing\n",
      "                       p-values.                                \u001b[35m[default: 999]\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-visualization\u001b[0m VISUALIZATION\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH    Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m  Display verbose output to stdout and/or stderr during\n",
      "                       execution of this action. Or silence output if\n",
      "                       execution is successful (silence is golden).\n",
      "  \u001b[94m--example-data\u001b[0m PATH  Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m          Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m               Show this message and exit.\n",
      "\n",
      "\u001b[33m                  There were some problems with the command:                  \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/2) Invalid value for '--i-distance-matrix': ../output/blank-mock/iversity-\n",
      "  core-bm/unweighted_unifrac_distance_matrix.qza does not exist.\u001b[0m\n",
      "\u001b[31m\u001b[1m (2/2) Missing option '--m-metadata-file'.\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/blank-mock/diversity-core-bm/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-slim.tsv \\\n",
    "  --o-visualization ../output/blank-mock/diversity-core-bm/unweighted-unifrac-body-site-significance.qzv \\\n",
    "  --p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/blank-mock/iversity-core-bm/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-slim.tsv \\\n",
    "  --o-visualization ../output/blank-mock/diversity-core-bm/unweighted-unifrac-body-site-significance.qzv \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e9a30b-96cf-4455-b040-2c21e93fd171",
   "metadata": {},
   "source": [
    "## [Atacama Soil Microbiome: Questions to Guide Data Analysis](https://docs.qiime2.org/2023.5/tutorials/atacama-soils/#paired-end-read-analysis-commands:~:text=Questions%20to%20guide%20data%20analysis)\n",
    "What sample metadata or combinations of sample metadata are most strongly associated with the differences in microbial composition of the samples? Are these associations stronger with unweighted UniFrac or with Bray-Curtis? Based on what you know about these metrics, what does that difference suggest? For exploring associations between continuous metadata and sample composition, the commands qiime metadata distance-matrix in combination with qiime diversity mantel and qiime diversity bioenv will be useful. These were not covered in the Moving Pictures tutorial, but you can learn about them by running them with the `--help` parameter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8380a16-f76f-4232-8f6c-aa99e14abf8b",
   "metadata": {},
   "source": [
    "## Alpha rarefaction plotting\n",
    "In this section we’ll explore alpha diversity as a function of sampling depth using the qiime diversity alpha-rarefaction visualizer. This visualizer computes one or more alpha diversity metrics at multiple sampling depths, in steps between 1 (optionally controlled with --p-min-depth) and the value provided as --p-max-depth. At each sampling depth step, 10 rarefied tables will be generated, and the diversity metrics will be computed for all samples in the tables. The number of iterations (rarefied tables computed at each sampling depth) can be controlled with --p-iterations. Average diversity values will be plotted for each sample at each even sampling depth, and samples can be grouped based on metadata in the resulting visualization if sample metadata is provided with the --m-metadata-file parameter.\n",
    "\n",
    "The value that you provide for --p-max-depth should be determined by reviewing the “Frequency per sample” information presented in the table.qzv file that was created above. In general, choosing a value that is somewhere around the median frequency seems to work well, but you may want to increase that value if the lines in the resulting rarefaction plot don’t appear to be leveling out, or decrease that value if you seem to be losing many of your samples due to low total frequencies closer to the minimum sampling depth than the maximum sampling depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "dfcb4d2b-a044-4434-9980-37282f97c045",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/alpha-rarefaction-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-rarefaction \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-phylogeny ../output/blank-mock/phylogeny-tree-bm/rooted_tree.qza \\\n",
    "  --p-max-depth 104183 \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/blank-mock/alpha-rarefaction-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c2a1e0bd-d356-4c77-a56e-3363864f7d86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/10bb300b-3e9b-43c9-b30c-76db405d1af8')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 10bb300b-3e9b-43c9-b30c-76db405d1af8>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/alpha-rarefaction-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6218ec4e-f66a-428e-ab2e-16bfe1d77144",
   "metadata": {},
   "source": [
    "## [Taxonomic Analysis](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#taxonomic-analysis:~:text=and%20bottom%20plots.)\n",
    "\n",
    "In the next sections we’ll begin to explore the taxonomic composition of the samples, and again relate that to sample metadata. The first step in this process is to assign taxonomy to the sequences in our FeatureData[Sequence] QIIME 2 artifact. We’ll do that using a pre-trained Naive Bayes classifier and the q2-feature-classifier plugin. This classifier was trained on the Greengenes 13_8 99% OTUs, where the sequences have been trimmed to only include 250 bases from the region of the 16S that was sequenced in this analysis (the V4 region, bound by the 515F/806R primer pair). We’ll apply this classifier to our sequences, and we can generate a visualization of the resulting mapping from sequence to taxonomy.\n",
    "\n",
    "Taxonomic classifiers perform best when they are trained based on your specific sample preparation and sequencing parameters, including the primers that were used for amplification and the length of your sequence reads. Therefore in general you should follow the instructions in Training feature classifiers with q2-feature-classifier to train your own taxonomic classifiers. We provide some common classifiers on our data resources page, including Silva-based 16S classifiers, though in the future we may stop providing these in favor of having users train their own classifiers which will be most relevant to their sequence data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "f9af4858-8be7-4279-b291-f76616e3229f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-08-16 12:05:51--  https://docs.qiime2.org/2021.11/data/tutorials/moving-pictures-usage/gg-13-8-99-515-806-nb-classifier.qza\n",
      "Resolving docs.qiime2.org (docs.qiime2.org)... 104.21.84.49, 172.67.186.144, 2606:4700:3035::ac43:ba90, ...\n",
      "Connecting to docs.qiime2.org (docs.qiime2.org)|104.21.84.49|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 28289645 (27M) [application/octet-stream]\n",
      "Saving to: ‘gg-13-8-99-515-806-nb-classifier.qza’\n",
      "\n",
      "gg-13-8-99-515-806- 100%[===================>]  26.98M  96.0MB/s    in 0.3s    \n",
      "\n",
      "2023-08-16 12:05:51 (96.0 MB/s) - ‘gg-13-8-99-515-806-nb-classifier.qza’ saved [28289645/28289645]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!cd ../output/blank-mock/taxonomy ; wget \\\n",
    "  -O 'gg-13-8-99-515-806-nb-classifier.qza' \\\n",
    "  'https://docs.qiime2.org/2021.11/data/tutorials/moving-pictures-usage/gg-13-8-99-515-806-nb-classifier.qza'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c62d59d6-f23d-498a-a958-43e04b063252",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[Taxonomy] to: ../output/blank-mock/taxonomy-bm.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: ../output/blank-mock/taxonomy-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier classify-sklearn \\\n",
    "  --i-classifier ../output/blank-mock/taxonomy/gg-13-8-99-515-806-nb-classifier.qza \\\n",
    "  --i-reads ../output/blank-mock/rep-seqs-bm.qza \\\n",
    "  --o-classification ../output/blank-mock/taxonomy-bm.qza\n",
    "!qiime metadata tabulate \\\n",
    "  --m-input-file ../output/blank-mock/taxonomy-bm.qza \\\n",
    "  --o-visualization ../output/blank-mock/taxonomy-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "989b33d4-b287-4b4b-94b8-5e4748be7899",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/e368638a-9d1f-493d-b344-61fc1e849780')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: e368638a-9d1f-493d-b344-61fc1e849780>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/taxonomy-bm.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "90b5928a-8f17-417f-b6ef-519917853684",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/taxa-bar-plots-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime taxa barplot \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-taxonomy ../output/blank-mock/taxonomy-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --o-visualization ../output/blank-mock/taxa-bar-plots-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c62a8364-855b-4e13-bb9e-235f2486cc4b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/583653c3-cb51-49d9-93fb-b9f80899096f')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 583653c3-cb51-49d9-93fb-b9f80899096f>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/taxa-bar-plots-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13bec273-d452-47ca-8830-0b189461c767",
   "metadata": {},
   "source": [
    "## [Differential Abundance Testing](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#taxonomic-analysis:~:text=the%20later%20timepoints%3F-,Differential%20abundance%20testing,-with%20ANCOM%C2%B6)\n",
    "\n",
    "ANCOM can be applied to identify features that are differentially abundant (i.e. present in different abundances) across sample groups. As with any bioinformatics method, you should be aware of the assumptions and limitations of ANCOM before using it. We recommend reviewing the ANCOM paper before using this method.\n",
    "\n",
    "Differential abundance testing in microbiome analysis is an active area of research. There are two QIIME 2 plugins that can be used for this: q2-gneiss and q2-composition. This section uses q2-composition, but there is another tutorial which uses gneiss on a different dataset if you are interested in learning more.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd51ed4b-68a3-4aec-b680-120429b7992e",
   "metadata": {},
   "source": [
    "### ANCOM\n",
    "**Analysis of Composition of Microbiomes**\n",
    "ANCOM is implemented in the q2-composition plugin. ANCOM assumes that few (less than about 25%) of the features are changing between groups. If you expect that more features are changing between your groups, you should not use ANCOM as it will be more error-prone (an increase in both Type I and Type II errors is possible). \r\n",
    "Here I do not filter any samples and use everything so I can see what is in the mock and blank samples. So we'll use the full feature table located at `../output/blank-mock/table-bm.qza`\n",
    "\n",
    "ANCOM operates on a FeatureTable[Composition] QIIME 2 artifact, which is based on frequencies of features on a per-sample basis, but cannot tolerate frequencies of zero. To build the composition artifact, a FeatureTable[Frequency] artifact must be provided to add-pseudocount (an imputation method), which will produce the FeatureTable[Composition] artifact.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a464fd7c-8b5d-44c5-a603-86077f376b24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Composition] to: ../output/blank-mock/comp-table-bm.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition add-pseudocount \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --o-composition-table ../output/blank-mock/comp-table-bm.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "557d65b9-54dc-48c6-aa4c-344d28d84fa8",
   "metadata": {},
   "source": [
    "We can then run ANCOM on the `PaeTemp` column to determine what features differ in abundance across the coral samples of the treatments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "59e53a9d-7c53-4c64-8130-29438777f0fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/ancom-paetemp-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition ancom \\\n",
    "  --i-table ../output/blank-mock/comp-table-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column PaeTemp \\\n",
    "  --o-visualization ../output/blank-mock/ancom-paetemp-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b8584eb-84cf-4992-af15-990e6421c50d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/c59a2c0c-a88a-44d8-8b74-eae794316f6b')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: c59a2c0c-a88a-44d8-8b74-eae794316f6b>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/ancom-paetemp-bm.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8514c98c-b093-4479-b230-a8ead940a315",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/blank-mock/ancom-temp-bm.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime composition ancom \\\n",
    "  --i-table ../output/blank-mock/comp-table-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/blank-mock/ancom-temp-bm.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "75210f4d-cf5e-47b7-8867-70855b7f348d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/041071ba-bfb3-40a2-b4bf-9e4830a1df69')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 041071ba-bfb3-40a2-b4bf-9e4830a1df69>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/ancom-temp-bm.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf4e91aa-17a6-40d6-ae2f-a6ff72a0d443",
   "metadata": {},
   "source": [
    "#### Taxonomy to Species (Level 7)\n",
    "\n",
    "We’re also often interested in performing a differential abundance test at a specific taxonomic level. To do this, we can collapse the features in our FeatureTable[Frequency] at the taxonomic level of interest, and then re-run the above steps. In this tutorial, we collapse our feature table at the genus level (i.e. level 6 of the Greengenes taxonomy).\n",
    "\n",
    "Level:\n",
    "   1. Kingdom\n",
    "   2. Phylum\n",
    "   3. Class\n",
    "   4. Order\n",
    "   5. Family\n",
    "   6. Genus\n",
    "   7. Species\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3611036b-82e9-4a16-aa07-b291839fa3f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: ../output/blank-mock/table-bm-l7.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: ../output/blank-mock/comp-table-bm-l7.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: ../output/blank-mock/ancom-temp-bm-l7.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime taxa collapse \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-taxonomy ../output/blank-mock/taxonomy-bm.qza \\\n",
    "  --p-level 7 \\\n",
    "  --o-collapsed-table ../output/blank-mock/table-bm-l7.qza\n",
    "!qiime composition add-pseudocount \\\n",
    "  --i-table ../output/blank-mock/table-bm-l7.qza \\\n",
    "  --o-composition-table ../output/blank-mock/comp-table-bm-l7.qza\n",
    "!qiime composition ancom \\\n",
    "  --i-table ../output/blank-mock/comp-table-bm-l7.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/blank-mock/ancom-temp-bm-l7.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "151a160a-3b44-40b0-aee5-d73c8def84e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/489771ad-fa30-4667-a44e-a2d8cbf8a860')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 489771ad-fa30-4667-a44e-a2d8cbf8a860>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/ancom-temp-bm-l7.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfaf0008-8e1c-48a4-b2a7-ed5c78218b2c",
   "metadata": {},
   "source": [
    "### [Gneiss](https://docs.qiime2.org/2023.5/tutorials/gneiss/#:~:text=Differential%20abundance%20analysis%20with%20gneiss,-%C2%B6)\n",
    "\n",
    "In this tutorial you will learn how to perform differential abundance analysis using balances in gneiss. The main problem that we will focus on is how to identify differentially abundant taxa in a compositionally coherent way.\r\n",
    "\r\n",
    "Compositionality refers to the issue of dealing with proportions. To account for differences in sequencing depth, microbial abundances are typically interpreted as proportions (e.g. relative abundance). Because of this, it becomes challenging to infer exactly which microbes are changing – since proportions add to one, the change of a single microbe will also change the proportions of the remaining microbes.\n",
    "Rather than focusing on individual taxa, we can focus on the ratio between taxa (or groups of taxa), since these ratios are consist between the true abundances and the observed proportions of the species observed. We typically log transform these ratios for improved visualization (‘log ratios’). The concept of calculating balances (or ratios) for multiple species can be extended to trees as shown in the following example.\n",
    "![balances taxon trees](https://docs.qiime2.org/2023.5/_images/gneiss-balances.jpg)\n",
    "On the left, we define a tree, where each of the tips corresponds to a taxon, and underneath are the proportions of each taxon in the first sample. The internal nodes (i.e. balances) define the log ratio between the taxa underneath. On the right is the same tree, and underneath are the proportions of each taxa in a different sample. Only one of the taxa abundances changes. As we have observed before, the proportions of all of the taxa will change, but looking at the balances, only the balances containing the purple taxa will change. In this case, balance b3\r\n",
    " won’t change, since it only considers the ratio between the red and taxa. By looking at balances instead proportions, we can eliminate some of the variance by restricting observations to only focus on the taxa within a given balance.\n",
    "\n",
    "The outstanding question here is, how do we construct a balance tree to control for the variation, and identify interesting differentially abundant partitions of taxa? In gneiss, there are three main ways that this can be done:\r\n",
    "1. \r\n",
    "Correlation clustering. If we don’t have relevant prior information about how to cluster together organisms, we can group together organisms based on how often they co-occur with each other. This is available in the correlation-clustering command and creates tree input for ilr-hierarchical.2. \r\n",
    "\r\n",
    "Gradient clustering. Use a metadata category to cluster taxa found in similar sample types. For example, if we want to evaluate if pH is a driving factor, we can cluster according to the pH that the taxa are observed in, and observe whether the ratios of low-pH organisms to high-pH organisms change as the pH changes. This is available in the gradient-clustering command and creates tree input for ilr-hierarchica3. l.\r\n",
    "\r\n",
    "Phylogenetic analysis. A phylogenetic tree (e.g. rooted-tree.qza) created outside of gneiss can also be used. In this case you can use your phylogenetic tree as input for ilr-phyl\n",
    "\n",
    "Once we have a tree, we can calculate balances...espectively.\r\n",
    "\r\n",
    "After the balances are calculated, standard statistical procedures such as ANOVA and linear regression ca\n",
    "\n",
    "First, we will define partitions of microbes for which we want to construct balances. Again, there are multiple possible ways to construct a tree (i.e. hierarchy) which defines the partition of microbes (balances) for which we want to construct balances. We will show examples of both correlation-clustering and gradient-clustering on this dataset.\r\n",
    "\r\n",
    "Note that the differential abundance techniques that we will be running will utilize log ratio transforms. Since it is not possible to take the logarithm of zero, both clustering methods below include a default pseudocount parameter. This replaces all zeroes in the table with a 1, so that we can apply logarithms on this transformed table.\r\n",
    "\r\n",
    "The input table is the raw count table (FeatureTable[Frequency]).n be performed.ogenetic."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51284fda-a2d0-4a87-a257-63e7f06629b7",
   "metadata": {},
   "source": [
    "#### Correlation Clustering\n",
    "This option should be your default option. We will employ unsupervised clustering via Ward’s hierarchical clustering to obtain Principal Balances. In essence, this will define the partitions of microbes that commonly co-occur with each other using Ward hierarchical clustering, which is defined by the following metric:\n",
    "\n",
    "$$\n",
    "d(x,y) = V[ln \\frac{x}{y}]\n",
    "$$\n",
    "\n",
    "Where **x**  and** ** \n",
    " represent the proportions of two microbes across all of the samples. If two microbes are highly correlated, then this quantity will shrink close to zero. Ward hierarchical cluster will then use this distance metric to iteratively cluster together groups of microbes that are correlated with each other. In the end, the tree that we obtain will highlight the high level structure and identify any blocks within in the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a7175159-5daa-439f-a4af-c6316252c98c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Hierarchy to: ../output/blank-mock/hierarchy-bm.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss correlation-clustering \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --o-clustering ../output/blank-mock/hierarchy-bm.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d84db6b6-6a55-4270-b4c9-db72598ef24c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m\u001b[1mPlugin warning from gneiss:\n",
      "\n",
      "dendrogram-heatmap is deprecated and will be removed in a future version of this plugin.\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/heatmap-cc-paetemp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss dendrogram-heatmap \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-tree ../output/blank-mock/hierarchy-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column PaeTemp \\\n",
    "  --p-color-map seismic \\\n",
    "  --o-visualization ../output/blank-mock/heatmap-cc-paetemp.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad5b21cc-c754-42aa-ae0e-dab256bffd98",
   "metadata": {},
   "source": [
    "##### PaeTemp Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "616012fc-b900-43c2-8b8b-b4d7beaa66fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/44b39522-a9bd-4972-8966-1e9e47926c77')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 44b39522-a9bd-4972-8966-1e9e47926c77>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/heatmap-cc-paetemp.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6f77014-6f24-4d41-9a89-e6ea80f9346a",
   "metadata": {},
   "source": [
    "##### Temp Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b4b77def-ab5c-4af3-9413-dff4eae0b504",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m\u001b[1mPlugin warning from gneiss:\n",
      "\n",
      "dendrogram-heatmap is deprecated and will be removed in a future version of this plugin.\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/heatmap-cc-temp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss dendrogram-heatmap \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-tree ../output/blank-mock/hierarchy-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --p-color-map seismic \\\n",
    "  --o-visualization ../output/blank-mock/heatmap-cc-temp.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b80f19ab-906b-4181-b2d4-4d41c15013e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/10fe0e0e-61a6-4ff3-80aa-09da99140eeb')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 10fe0e0e-61a6-4ff3-80aa-09da99140eeb>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/heatmap-cc-temp.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6337c57c-fb20-45d6-ad0a-c6c8621f9898",
   "metadata": {},
   "source": [
    "#### Gradient Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "576ac1ef-bfe0-427c-947c-7fa69b0d7f97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime gneiss gradient-clustering\u001b[0m [OPTIONS]\n",
      "\n",
      "  Build a bifurcating tree that represents a hierarchical clustering of\n",
      "  features.  The hiearchical clustering uses Ward hierarchical clustering\n",
      "  based on the mean difference of gradients that each feature is observed in.\n",
      "  This method is primarily used to sort the table to reveal the underlying\n",
      "  block-like structures.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency | RelativeFrequency |\u001b[0m\n",
      "    \u001b[32mComposition]\u001b[0m       The feature table containing the samples in which the\n",
      "                       columns will be clustered.                   \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--m-gradient-file\u001b[0m METADATA\n",
      "  \u001b[94m\u001b[4m--m-gradient-column\u001b[0m COLUMN  \u001b[32mMetadataColumn[Numeric]\u001b[0m\n",
      "                       Contains gradient values to sort the features and\n",
      "                       samples.                                     \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--p-ignore-missing-samples\u001b[0m / \u001b[94m--p-no-ignore-missing-samples\u001b[0m\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-weighted\u001b[0m / \u001b[94m--p-no-weighted\u001b[0m\n",
      "                       Specifies if abundance or presence/absence information\n",
      "                       should be used to perform the clustering.\n",
      "                                                               \u001b[35m[default: True]\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-clustering\u001b[0m ARTIFACT\n",
      "    \u001b[32mHierarchy\u001b[0m          A hierarchy of feature identifiers where each tip\n",
      "                       corresponds to the feature identifiers in the table.\n",
      "                       This tree can contain tip ids that are not present in\n",
      "                       the table, but all feature ids in the table must be\n",
      "                       present in this tree.                        \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH    Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m  Display verbose output to stdout and/or stderr during\n",
      "                       execution of this action. Or silence output if\n",
      "                       execution is successful (silence is golden).\n",
      "  \u001b[94m--example-data\u001b[0m PATH  Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m          Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m               Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--m-gradient-file': Metadata column is of type\n",
      "  'categorical', but expected Numeric.\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss gradient-clustering \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --m-gradient-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-gradient-column PaeTemp \\\n",
    "  --o-clustering ../output/blank-mock/gradient-hierarchy-bm.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39fccdfb-c73f-4431-82c8-02033816a6b7",
   "metadata": {},
   "source": [
    "#### Phylogenetic Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e2bcaca8-9274-4929-9cc3-4edb3bf6f58f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime gneiss\u001b[0m [OPTIONS] COMMAND [ARGS]...\n",
      "\n",
      "  Description: This is a QIIME 2 plugin supporting statistical models on\n",
      "  feature tables and metadata using balances.\n",
      "\n",
      "  Plugin website: https://biocore.github.io/gneiss/\n",
      "\n",
      "  Getting user support: Please post to the QIIME 2 forum for help with this\n",
      "  plugin: https://forum.qiime2.org\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m--version\u001b[0m            Show the version and exit.\n",
      "  \u001b[94m--example-data\u001b[0m PATH  Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m          Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m               Show this message and exit.\n",
      "\n",
      "\u001b[1mCommands\u001b[0m:\n",
      "  \u001b[94massign-ids\u001b[0m                     Assigns ids on internal nodes in the tree,\n",
      "                                 and makes sure that they are consistent with\n",
      "                                 the table columns.\n",
      "  \u001b[94mcorrelation-clustering\u001b[0m         Hierarchical clustering using feature\n",
      "                                 correlation.\n",
      "  \u001b[94mdendrogram-heatmap\u001b[0m             Dendrogram heatmap.\n",
      "  \u001b[94mgradient-clustering\u001b[0m            Hierarchical clustering using gradient\n",
      "                                 information.\n",
      "  \u001b[94milr-hierarchical\u001b[0m               Isometric Log-ratio Transform applied to a\n",
      "                                 hierarchical clustering\n",
      "  \u001b[94milr-phylogenetic\u001b[0m               Isometric Log-ratio Transform applied to a\n",
      "                                 phylogenetic tree\n",
      "  \u001b[94milr-phylogenetic-differential\u001b[0m  Differentially abundant Phylogenetic Log\n",
      "                                 Ratios.\n",
      "  \u001b[94milr-phylogenetic-ordination\u001b[0m    Ordination through a phylogenetic Isometric\n",
      "                                 Log Ratio transform.\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "16df09c3-1184-4e5f-a23a-f480dac2e2f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Balance] to: ../output/blank-mock/balances-rootedtree-bm.qza\u001b[0m\n",
      "\u001b[32mSaved Hierarchy to: ../output/blank-mock/hierarchy-rootedtree-bm.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss ilr-phylogenetic \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-tree ../output/blank-mock/phylogeny-tree-bm/rooted_tree.qza \\\n",
    "  --o-balances ../output/blank-mock/balances-rootedtree-bm.qza \\\n",
    "  --o-hierarchy ../output/blank-mock/hierarchy-rootedtree-bm.qza\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "720624bd-fcdf-4cd3-968b-341fc34f5a28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m\u001b[1mPlugin warning from gneiss:\n",
      "\n",
      "dendrogram-heatmap is deprecated and will be removed in a future version of this plugin.\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/blank-mock/heatmap-rootedtree-temp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime gneiss dendrogram-heatmap \\\n",
    "  --i-table ../output/blank-mock/table-bm.qza \\\n",
    "  --i-tree ../output/blank-mock/hierarchy-rootedtree-bm.qza \\\n",
    "  --m-metadata-file ../rawdata/sample-metadata-verbose.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --p-color-map seismic \\\n",
    "  --o-visualization ../output/blank-mock/heatmap-rootedtree-temp.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "554215fe-a0b1-44e1-b6d0-51ad49302ad8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/f6b50e5d-595d-4cf7-a266-d72cd6d31623')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: f6b50e5d-595d-4cf7-a266-d72cd6d31623>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/blank-mock/heatmap-rootedtree-temp.qzv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
