{
 "cells": [
  {
   "cell_type": "raw",
   "id": "e15ca787-fd40-4888-94e7-b78a721b05b9",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Microbiome Sequence Analysis\"\n",
    "subtitle: \"Community composition from 16S Sequence Data of Montipora capitata corals exposed to heat and plasticizers\"\n",
    "author: \"Sarah Tanja\"\n",
    "format: html\n",
    "editor: visual\n",
    "toc: true\n",
    "toc-title: Contents <i class=\"bi bi-bookmark-heart\"></i>\n",
    "toc-depth: 5\n",
    "toc-location: left\n",
    "reference-location: margin\n",
    "citation-location: margin\n",
    "bibliography: ./references/qiime2.bib\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c12bb8-df65-4c49-b01a-11d3b1b8b664",
   "metadata": {},
   "source": [
    "# Amplicon Sequence Data Analysis with QIIME 2\n",
    "\n",
    "Adding ```!``` before the command tells the notebook this is a bash command, rather than python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e670c1b-4199-4cdf-9fa3-27c433bfcf6c",
   "metadata": {},
   "source": [
    "To use sequencing data in QIIME2, we first need to turn the FASTQ files containing our data into QIIME artifacts.\n",
    "\n",
    "What the QIIME2 pipeline will do:\r\n",
    "![our workflow](https://github.com/Gibbons-Lab/isb_course_2023/raw/main/docs/16S/assets/steps.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "435a4d8e-dc7f-4263-8e42-033506a1dd2d",
   "metadata": {},
   "source": [
    "### About the Data\n",
    "I downloaded FASTQ data files generated by [Mr. DNA Lab](https://www.mrdnalab.com/) Molecular Research from DropBox. \n",
    "I unzipped the folders and uploaded the `sample-metadata.tsv` file and `demux` folder into the `coral-pae-temp/analysis/microbiome/data` directory.  \n",
    "\n",
    "Here I am working with the FASTQ data files located in `coral-pae-temp/analysis/microbiome/data/demux`. In the `demux` folder are two `fastq.gz` files for each of the 22 samples, one for the forward read and one for the reverse read. \n",
    "\n",
    "The `fastq.gz` file name includes the sample identifier and should look like `4.Ea_S1_L001_R1_001.fastq.gz`. \n",
    "The underscore-separated fields in this file name are:\n",
    "\n",
    "1.  the sample identifier,\n",
    "\n",
    "2.  the barcode sequence or a barcode identifier,\n",
    "\n",
    "3.  the lane number,\n",
    "\n",
    "4.  the direction of the read (i.e. R1 or R2, because these are paired-end reads), and\n",
    "\n",
    "5.  the set number.\n",
    "   \n",
    "\n",
    "The `fastq.gz` files are **Demultiplexed** (aka **Demuxed**) sequences that still have the forward and reverse primers in the sequences.\n",
    "\n",
    "-   The Raw Data is **demultiplexed**\n",
    "\n",
    "-   A R1 and R2 fastq.gz file has been generated for each individual sample\n",
    "\n",
    "-   All forward reads are binned into the R1 fastq.gz files\n",
    "\n",
    "-   All reverse reads are binned into the R2 fastq.gz files\n",
    "\n",
    "-   Other than demultiplexing; you can consider the Raw Data on BaseSpace as untouched (**The Forward and Reverse Primer Sequences have not been removed**)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7997d9-10ca-4323-85e2-676ee96f60b5",
   "metadata": {},
   "source": [
    "## Python 3 API import qiime plugins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a593cfd4-fcb7-4a42-8440-457f99e65015",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1cbf68be-5a29-42a6-bf59-780ff68a095c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install empress\n",
    "#!qiime dev refresh-cache\n",
    "#!qiime empress --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aac072b0-828e-4ce3-bdd7-93e548318b80",
   "metadata": {},
   "source": [
    "### Treatment\n",
    "Only treatment samples (remove Environmental control samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c476a61f-a6c6-4b18-8ec4-782fa8e0297c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "25b5f1d2-78ff-417c-9a60-95a81f72c188",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>BarcodeName</th>\n",
       "      <th>ReversePrimer</th>\n",
       "      <th>ProjectName</th>\n",
       "      <th>Description</th>\n",
       "      <th>Pae</th>\n",
       "      <th>Temp</th>\n",
       "      <th>PaeTemp</th>\n",
       "      <th>Colony</th>\n",
       "      <th>Tank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.CA2a</td>\n",
       "      <td>ATCATAGGCT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0073</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>1.CA2a</td>\n",
       "      <td>control</td>\n",
       "      <td>ambient</td>\n",
       "      <td>control-ambient</td>\n",
       "      <td>1</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.CH2a</td>\n",
       "      <td>TGTTAGAAGG</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0074</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>1.CH2a</td>\n",
       "      <td>control</td>\n",
       "      <td>hot</td>\n",
       "      <td>control-hot</td>\n",
       "      <td>1</td>\n",
       "      <td>H2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.PA2a</td>\n",
       "      <td>ACGGCCGTCA</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0076</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>1.PA2a</td>\n",
       "      <td>peak</td>\n",
       "      <td>ambient</td>\n",
       "      <td>peak-ambient</td>\n",
       "      <td>1</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.PH1a</td>\n",
       "      <td>CGTTGCTTAC</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0077</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>1.PH1a</td>\n",
       "      <td>peak</td>\n",
       "      <td>hot</td>\n",
       "      <td>peak-hot</td>\n",
       "      <td>1</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.CA2a</td>\n",
       "      <td>TGACTACATA</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0078</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>2.CA2a</td>\n",
       "      <td>control</td>\n",
       "      <td>ambient</td>\n",
       "      <td>control-ambient</td>\n",
       "      <td>2</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.CH1b</td>\n",
       "      <td>CGGCCTCGTT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0079</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>2.CH1b</td>\n",
       "      <td>control</td>\n",
       "      <td>hot</td>\n",
       "      <td>control-hot</td>\n",
       "      <td>2</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2.PA1b</td>\n",
       "      <td>TCGTCTGACT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0081</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>2.PA1b</td>\n",
       "      <td>peak</td>\n",
       "      <td>ambient</td>\n",
       "      <td>peak-ambient</td>\n",
       "      <td>2</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.PH2a</td>\n",
       "      <td>CTCATAGCGA</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0082</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>2.PH2a</td>\n",
       "      <td>peak</td>\n",
       "      <td>hot</td>\n",
       "      <td>peak-hot</td>\n",
       "      <td>2</td>\n",
       "      <td>H2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3.CA1b</td>\n",
       "      <td>AGACACATTA</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0083</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>3.CA1b</td>\n",
       "      <td>control</td>\n",
       "      <td>ambient</td>\n",
       "      <td>control-ambient</td>\n",
       "      <td>3</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3.CH2a</td>\n",
       "      <td>GCGCGATGTT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0084</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>3.CH2a</td>\n",
       "      <td>control</td>\n",
       "      <td>hot</td>\n",
       "      <td>control-hot</td>\n",
       "      <td>3</td>\n",
       "      <td>H2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3.PA1a</td>\n",
       "      <td>ACGTCAATAC</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0086</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>3.PA1a</td>\n",
       "      <td>peak</td>\n",
       "      <td>ambient</td>\n",
       "      <td>peak-ambient</td>\n",
       "      <td>3</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3.PH1b</td>\n",
       "      <td>GATACCTCCT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0087</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>3.PH1b</td>\n",
       "      <td>peak</td>\n",
       "      <td>hot</td>\n",
       "      <td>peak-hot</td>\n",
       "      <td>3</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4.CA1a</td>\n",
       "      <td>ATCCGTAAGT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0088</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>4.CA1a</td>\n",
       "      <td>control</td>\n",
       "      <td>ambient</td>\n",
       "      <td>control-ambient</td>\n",
       "      <td>4</td>\n",
       "      <td>A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4.CH1a</td>\n",
       "      <td>CGTGTATCTT</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0089</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>4.CH1a</td>\n",
       "      <td>control</td>\n",
       "      <td>hot</td>\n",
       "      <td>control-hot</td>\n",
       "      <td>4</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>4.PA2a</td>\n",
       "      <td>GGCCATCATA</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0091</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>4.PA2a</td>\n",
       "      <td>peak</td>\n",
       "      <td>ambient</td>\n",
       "      <td>peak-ambient</td>\n",
       "      <td>4</td>\n",
       "      <td>A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>4.PH1b</td>\n",
       "      <td>ACATACTTCC</td>\n",
       "      <td>GTGYCAGCMGCCGCGGTAA</td>\n",
       "      <td>60bp_UDPi5_0092</td>\n",
       "      <td>GGACTACNVGGGTWTCTAAT</td>\n",
       "      <td>060823STillcus515F</td>\n",
       "      <td>4.PH1b</td>\n",
       "      <td>peak</td>\n",
       "      <td>hot</td>\n",
       "      <td>peak-hot</td>\n",
       "      <td>4</td>\n",
       "      <td>H1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #SampleID BarcodeSequence LinkerPrimerSequence      BarcodeName  \\\n",
       "0     1.CA2a      ATCATAGGCT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0073   \n",
       "1     1.CH2a      TGTTAGAAGG  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0074   \n",
       "3     1.PA2a      ACGGCCGTCA  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0076   \n",
       "4     1.PH1a      CGTTGCTTAC  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0077   \n",
       "5     2.CA2a      TGACTACATA  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0078   \n",
       "6     2.CH1b      CGGCCTCGTT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0079   \n",
       "8     2.PA1b      TCGTCTGACT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0081   \n",
       "9     2.PH2a      CTCATAGCGA  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0082   \n",
       "10    3.CA1b      AGACACATTA  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0083   \n",
       "11    3.CH2a      GCGCGATGTT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0084   \n",
       "13    3.PA1a      ACGTCAATAC  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0086   \n",
       "14    3.PH1b      GATACCTCCT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0087   \n",
       "15    4.CA1a      ATCCGTAAGT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0088   \n",
       "16    4.CH1a      CGTGTATCTT  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0089   \n",
       "18    4.PA2a      GGCCATCATA  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0091   \n",
       "19    4.PH1b      ACATACTTCC  GTGYCAGCMGCCGCGGTAA  60bp_UDPi5_0092   \n",
       "\n",
       "           ReversePrimer         ProjectName Description       Pae     Temp  \\\n",
       "0   GGACTACNVGGGTWTCTAAT  060823STillcus515F      1.CA2a   control  ambient   \n",
       "1   GGACTACNVGGGTWTCTAAT  060823STillcus515F      1.CH2a   control      hot   \n",
       "3   GGACTACNVGGGTWTCTAAT  060823STillcus515F      1.PA2a      peak  ambient   \n",
       "4   GGACTACNVGGGTWTCTAAT  060823STillcus515F      1.PH1a      peak      hot   \n",
       "5   GGACTACNVGGGTWTCTAAT  060823STillcus515F      2.CA2a   control  ambient   \n",
       "6   GGACTACNVGGGTWTCTAAT  060823STillcus515F      2.CH1b  control       hot   \n",
       "8   GGACTACNVGGGTWTCTAAT  060823STillcus515F      2.PA1b      peak  ambient   \n",
       "9   GGACTACNVGGGTWTCTAAT  060823STillcus515F      2.PH2a      peak      hot   \n",
       "10  GGACTACNVGGGTWTCTAAT  060823STillcus515F      3.CA1b   control  ambient   \n",
       "11  GGACTACNVGGGTWTCTAAT  060823STillcus515F      3.CH2a   control      hot   \n",
       "13  GGACTACNVGGGTWTCTAAT  060823STillcus515F      3.PA1a      peak  ambient   \n",
       "14  GGACTACNVGGGTWTCTAAT  060823STillcus515F      3.PH1b      peak      hot   \n",
       "15  GGACTACNVGGGTWTCTAAT  060823STillcus515F      4.CA1a   control  ambient   \n",
       "16  GGACTACNVGGGTWTCTAAT  060823STillcus515F      4.CH1a   control      hot   \n",
       "18  GGACTACNVGGGTWTCTAAT  060823STillcus515F      4.PA2a      peak  ambient   \n",
       "19  GGACTACNVGGGTWTCTAAT  060823STillcus515F      4.PH1b      peak      hot   \n",
       "\n",
       "            PaeTemp Colony Tank  \n",
       "0   control-ambient      1   A2  \n",
       "1       control-hot      1   H2  \n",
       "3      peak-ambient      1   A2  \n",
       "4          peak-hot      1   H1  \n",
       "5   control-ambient      2   A2  \n",
       "6       control-hot      2   H1  \n",
       "8      peak-ambient      2   A1  \n",
       "9          peak-hot      2   H2  \n",
       "10  control-ambient      3   A1  \n",
       "11      control-hot      3   H2  \n",
       "13     peak-ambient      3   A1  \n",
       "14         peak-hot      3   H1  \n",
       "15  control-ambient      4   A1  \n",
       "16      control-hot      4   H1  \n",
       "18     peak-ambient      4   A2  \n",
       "19         peak-hot      4   H1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in tsv sample metadata as a csv\n",
    "df = pd.read_csv('../data/sample-metadata-verbose.tsv', delimiter='\\t')\n",
    "# make a list of SampleID values to remove\n",
    "remove = ['1.Ea', '2.Ea', '3.Eb', '4.Ea', 'blank.', 'mock.']\n",
    "# remove those rows \n",
    "df = df[~df['#SampleID'].isin(remove)]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9f81dd22-7694-4917-9d33-5da0d9a278e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the new df as a tsv\n",
    "df.to_csv('../data/sample-metadata-treatment.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa946f36-9902-4a9e-b0e3-adc4faf37a3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: ../output/filtered/table-treatment.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table filter-samples \\\n",
    "  --i-table ../output/filtered/table-no-mock-no-hits-taxon-filtered.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --o-filtered-table ../output/filtered/table-treatment.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "63eeccbc-545f-4696-b699-33d48d2122e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/filtered/table-treatment.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime feature-table summarize \\\n",
    "  --i-table ../output/filtered/table-treatment.qza \\\n",
    "  --o-visualization ../output/filtered/table-treatment.qzv \\\n",
    "  --m-sample-metadata-file ../data/sample-metadata-treatment.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f74905c7-7813-44bd-a27d-0350d9866e36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/a0cbf4dc-efc4-45a5-bf1a-cd4a8789cbec')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: a0cbf4dc-efc4-45a5-bf1a-cd4a8789cbec>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/filtered/table-treatment.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3249d3a-d377-4835-a186-17533dbf1190",
   "metadata": {},
   "source": [
    "## Diversity & Phylogenetics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5be80bf-079a-4aaa-951f-a29a68553acc",
   "metadata": {},
   "source": [
    "[Generate a tree for phylogenetic diversity analyses](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Generate%20a%20tree%20for%20phylogenetic%20diversity%20analyses)\n",
    "\n",
    "From the moving pictures tutorial:\n",
    "> QIIME supports several phylogenetic diversity metrics, including Faith’s Phylogenetic Diversity and weighted and unweighted UniFrac. In addition to counts of features per sample (i.e., the data in the FeatureTable[Frequency] QIIME 2 artifact), these metrics require a rooted phylogenetic tree relating the features to one another. This information will be stored in a Phylogeny[Rooted] QIIME 2 artifact. To generate a phylogenetic tree we will use align-to-tree-mafft-fasttree pipeline from the q2-phylogeny plugin. \r\n",
    "First, the pipeline uses the mafft program to perform a multiple sequence alignment of the sequences in our FeatureData[Sequence] to create a FeatureData[AlignedSequence] QIIME 2 artifact. Next, the pipeline masks (or filters) the alignment to remove positions that are highly variable. These positions are generally considered to add noise to a resulting phylogenetic tree. Following that, the pipeline applies FastTree to generate a phylogenetic tree from the masked alignment. The FastTree program creates an unrooted tree, so in the final step in this section midpoint rooting is applied to place the root of the tree at the midpoint of the longest tip-to-tip distance in the unrooted tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "17b6c955-67f3-4f29-b3e9-abbdaf5190dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime phylogeny align-to-tree-mafft-fasttree\u001b[0m [OPTIONS]\n",
      "\n",
      "  This pipeline will start by creating a sequence alignment using MAFFT, after\n",
      "  which any alignment columns that are phylogenetically uninformative or\n",
      "  ambiguously aligned will be removed (masked). The resulting masked alignment\n",
      "  will be used to infer a phylogenetic tree and then subsequently rooted at\n",
      "  its midpoint. Output files from each step of the pipeline will be saved.\n",
      "  This includes both the unmasked and masked MAFFT alignment from q2-alignment\n",
      "  methods, and both the rooted and unrooted phylogenies from q2-phylogeny\n",
      "  methods.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-sequences\u001b[0m ARTIFACT \u001b[32mFeatureData[Sequence]\u001b[0m\n",
      "                          The sequences to be used for creating a fasttree\n",
      "                          based rooted phylogenetic tree.           \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m--p-n-threads\u001b[0m VALUE \u001b[32mInt % Range(1, None) | Str % Choices('auto')\u001b[0m\n",
      "                          The number of threads. (Use `auto` to automatically\n",
      "                          use all available cores) This value is used when\n",
      "                          aligning the sequences and creating the tree with\n",
      "                          fasttree.                               \u001b[35m[default: 1]\u001b[0m\n",
      "  \u001b[94m--p-mask-max-gap-frequency\u001b[0m PROPORTION \u001b[32mRange(0, 1, inclusive_end=True)\u001b[0m\n",
      "                          The maximum relative frequency of gap characters in\n",
      "                          a column for the column to be retained. This\n",
      "                          relative frequency must be a number between 0.0 and\n",
      "                          1.0 (inclusive), where 0.0 retains only those\n",
      "                          columns without gap characters, and 1.0 retains all\n",
      "                          columns  regardless of gap character frequency. This\n",
      "                          value is used when masking the aligned sequences.\n",
      "                                                                \u001b[35m[default: 1.0]\u001b[0m\n",
      "  \u001b[94m--p-mask-min-conservation\u001b[0m PROPORTION \u001b[32mRange(0, 1, inclusive_end=True)\u001b[0m\n",
      "                          The minimum relative frequency of at least one\n",
      "                          non-gap character in a column for that column to be\n",
      "                          retained. This relative frequency must be a number\n",
      "                          between 0.0 and 1.0 (inclusive). For example, if a\n",
      "                          value of  0.4 is provided, a column will only be\n",
      "                          retained  if it contains at least one character that\n",
      "                          is present in at least 40% of the sequences. This\n",
      "                          value is used when masking the aligned sequences.\n",
      "                                                                \u001b[35m[default: 0.4]\u001b[0m\n",
      "  \u001b[94m--p-parttree\u001b[0m / \u001b[94m--p-no-parttree\u001b[0m\n",
      "                          This flag is required if the number of sequences\n",
      "                          being aligned are larger than 1000000. Disabled by\n",
      "                          default.                            \u001b[35m[default: False]\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-alignment\u001b[0m ARTIFACT \u001b[32mFeatureData[AlignedSequence]\u001b[0m\n",
      "                          The aligned sequences.                    \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-masked-alignment\u001b[0m ARTIFACT \u001b[32mFeatureData[AlignedSequence]\u001b[0m\n",
      "                          The masked alignment.                     \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-tree\u001b[0m ARTIFACT       The unrooted phylogenetic tree.\n",
      "    \u001b[32mPhylogeny[Unrooted]\u001b[0m                                             \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--o-rooted-tree\u001b[0m ARTIFACT\n",
      "    \u001b[32mPhylogeny[Rooted]\u001b[0m     The rooted phylogenetic tree.             \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH       Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m     Display verbose output to stdout and/or stderr\n",
      "                          during execution of this action. Or silence output\n",
      "                          if execution is successful (silence is golden).\n",
      "  \u001b[94m--recycle-pool\u001b[0m TEXT     Use a cache pool for pipeline resumption. QIIME 2\n",
      "                          will cache your results in this pool for reuse by\n",
      "                          future invocations. These pool are retained until\n",
      "                          deleted by the user. If not provided, QIIME 2 will\n",
      "                          create a pool which is automatically reused by\n",
      "                          invocations of the same action and removed if the\n",
      "                          action is successful. Note: these pools are local to\n",
      "                          the cache you are using.\n",
      "  \u001b[94m--no-recycle\u001b[0m            Do not recycle results from a previous failed\n",
      "                          pipeline run or save the results from this run for\n",
      "                          future recycling.\n",
      "  \u001b[94m--parallel\u001b[0m              Execute your action in parallel. This flag will use\n",
      "                          your default parallel config.\n",
      "  \u001b[94m--parallel-config\u001b[0m FILE  Execute your action in parallel using a config at\n",
      "                          the indicated path.\n",
      "  \u001b[94m--use-cache\u001b[0m DIRECTORY   Specify the cache to be used for the intermediate\n",
      "                          work of this pipeline. If not provided, the default\n",
      "                          cache under $TMP/qiime2/<uname> will be used.\n",
      "                          IMPORTANT FOR HPC USERS: If you are on an HPC system\n",
      "                          and are using parallel execution it is important to\n",
      "                          set this to a location that is globally accessible\n",
      "                          to all nodes in the cluster.\n",
      "  \u001b[94m--example-data\u001b[0m PATH     Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m             Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m                  Show this message and exit.\n",
      "\n",
      "\u001b[1mExamples\u001b[0m:\n",
      "  # ### example: align to tree mafft fasttree\n",
      "  qiime phylogeny align-to-tree-mafft-fasttree \\\n",
      "    --i-sequences rep-seqs.qza \\\n",
      "    --o-alignment aligned-rep-seqs.qza \\\n",
      "    --o-masked-alignment masked-aligned-rep-seqs.qza \\\n",
      "    --o-tree unrooted-tree.qza \\\n",
      "    --o-rooted-tree rooted-tree.qza\n",
      "  \n",
      "\n",
      "\u001b[33m                  There were some problems with the command:                  \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/5) Invalid value for '--output-dir': '../output/tree' already exists, will\n",
      "  not overwrite.\u001b[0m\n",
      "\u001b[31m\u001b[1m (2/5) Missing option '--o-alignment'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (3/5) Missing option '--o-masked-alignment'. (\"--output-dir\" may also be\n",
      "  used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (4/5) Missing option '--o-tree'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[31m\u001b[1m (5/5) Missing option '--o-rooted-tree'. (\"--output-dir\" may also be used)\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
    "  --i-sequences ../output/dada2/representative_sequences.qza \\\n",
    "  --output-dir ../output/tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "31d90c03-763a-4790-8de9-a6109d50fa5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/tree/empress.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime empress tree-plot \\\n",
    "    --i-tree ../output/tree/rooted_tree.qza \\\n",
    "    --o-visualization ../output/tree/empress.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f7e5e450-bd8e-450a-a104-ee659cb105df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/86ad3f28-2185-4d64-90fd-359fb2115295')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 86ad3f28-2185-4d64-90fd-359fb2115295>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/tree/empress.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d583891-a192-4fef-a7df-106b4a4a930c",
   "metadata": {},
   "source": [
    "[Diversity](https://docs.qiime2.org/2023.5/tutorials/moving-pictures-usage/#:~:text=Alpha%20and%20beta%20diversity%20analysis)\n",
    "\n",
    "QIIME 2’s diversity analyses are available through the `q2-diversity` plugin, which supports computing alpha and beta diversity metrics, applying related statistical tests, and generating interactive visualizations. We’ll first apply the `core-metrics-phylogenetic` method, which rarefies a `FeatureTable Frequency` to a user-specified depth, computes several alpha and beta diversity metrics, and generates principle coordinates analysis (PCoA) plots using Emperor for each of the beta diversity metrics.\n",
    "\n",
    "An important metric to consider when studying microbial ecology is __diversity__. Diversity comes in two flavors: ⍺ (alpha) and β (beta).\r\n",
    "\r\n",
    "Alpha diversity is pretty simple - how diverse is a single sample? You might consider measures like richness and evenness.\r\n",
    "\r\n",
    "![alpha diversity](https://gibbons-lab.github.io/isb_course_2023/16S/assets/alpha_diversity.png)\r\n",
    "\r\n",
    "Beta diversity instead looks at how different two samples are from each other - what taxa are shared, and how their abundances differ.\r\n",
    "\r\n",
    "![beta diversity](https://gibbons-lab.github.io/isb_course_2023/16S/assets/beta_diversity.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e514fb-bd7f-454c-acbc-1aa714df6144",
   "metadata": {},
   "source": [
    "The metrics computed by default are:\n",
    "\r",
    "##### Alpha diversity\r",
    "- \r\n",
    "Shannon’s diversity index (a quantitative measure of community richness- \r\n",
    "\r\n",
    "Observed Features (a qualitative measure of community richne- s)\r\n",
    "\r\n",
    "Faith’s Phylogenetic Dive**rsity (a qualitative measure of community richness that incorporates phylogenetic relationships between the fea**t- res)\r\n",
    "\r\n",
    "Evenness (or Pielou’s Evenness; a measure of community eve##### nness)\r\n",
    "\r\n",
    "Beta - iversity\r\n",
    "\r\n",
    "Jaccard distance (a qualitative measure of community dis- imilarity)\r\n",
    "\r\n",
    "Bray-Curtis distance (a quantitative measure of community d- ssimilarity)\r\n",
    "\r\n",
    "unweighted U**niFrac distance (a qualitative measure of community dissimilarity that incorporates phylogenetic relationships betwe**e-  the features)\r\n",
    "\r\n",
    "weighted** UniFrac distance (a quantitative measure of community dissimilarity that incorporates phylogenetic relationships bet**ween the features) \r"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a79db12-0620-4c43-a365-7e15ae264faa",
   "metadata": {},
   "source": [
    "### Sampling Depth\n",
    "\n",
    "An important parameter that needs to be provided to this code is ` --p-sampling-depth`, which is the even sampling (i.e. rarefaction) depth. Because most diversity metrics are sensitive to different sampling depths across different samples, this script will randomly subsample the counts from each sample to the value provided for this parameter. For am--ple, if you provide `--p-sampling-depth 500`, this step will subsample the counts in each sample without replacement so that each sample in the resulting table has a total count of 500. If the total count for any sample(s) are smaller than this value, those samples will be dropped from the diversity analysis. Choosing this value is tricky. We recommend making your choice by reviewing the information presented in the `table.qzv` file that was created above. Choose a value that is as high as possible (so you retain more sequences per sample) while excluding as few samples as possible."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3a1eaf6-d9aa-45f8-aab2-ff972d907c08",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "Navigate to the interactive sample detail tab\n",
    "<br>\n",
    "Move the sampling depth slider as high as you can before excluding any samples \n",
    "<br>\n",
    "We want the sampling depth to be high, while retaining all 22 samples\n",
    "<br>\n",
    "This looks like a sampling depth of 10,6727 (09AUG2023, SST) \n",
    "<br>\n",
    "But maybe we exclude the mock community and bring it up to 163971? (15AUG2023, SST)\n",
    "<br>\n",
    "... Redoing this with only the 16 samples that were exposed to the treatment ( so we're excluding here the mock and blank and environmental baseline samples)\n",
    "<br>\n",
    "In this.. the lowest frequency is 166920 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dafb35a5-8576-43a4-a82e-e73ff702e5ab",
   "metadata": {},
   "source": [
    "What value would you choose to pass for --p-sampling-depth? \n",
    "- **160000**\n",
    "How many samples will be excluded from your analysis based on this choice? \n",
    "- **none, all 16 treatment samples are retained**\n",
    "How many total sequences will you be analyzing in the core-metrics-phylogenetic command?\n",
    "Retained 2,560,000 (66.13%) features in 16 (100.00%) samples at the specifed sampling depth."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f41c70-cc79-4b30-8937-b8e0a6a1fbd9",
   "metadata": {},
   "source": [
    "To account for variations in sampling depth, we'll provide QIIME2 with a cutoff at which rarefy all our samples. Since this randomly selects sequences, your results might look a little different. We'll also pass in our metadata file, so we can keep track how which samples come from each group."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "251ed433-e0cc-4774-b724-99d8330c9749",
   "metadata": {},
   "source": [
    "## Alpha rarefaction plotting\n",
    "In this section we’ll explore alpha diversity as a function of sampling depth using the `qiime diversity alpha-rarefaction` visualizer. This visualizer computes one or more alpha diversity metrics at multiple sampling depths, in steps between 1 (optionally controlled with `--p-min-depth`) and the value provided as `--p-max-depth`. At each sampling depth step, 10 rarefied tables will be generated, and the diversity metrics will be computed for all samples in the tables. The number of iterations (rarefied tables computed at each sampling depth) can be controlled with `--p-iterations`. Average diversity values will be plotted for each sample at each even sampling depth, and samples can be grouped based on metadata in the resulting visualization if sample metadata is provided with the `--m-metadata-file` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dfcb4d2b-a044-4434-9980-37282f97c045",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/alpha-treatment/alpha-rarefaction.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-rarefaction \\\n",
    "  --i-table ../output/filtered/table-treatment.qza \\\n",
    "  --i-phylogeny ../output/tree/rooted_tree.qza \\\n",
    "  --p-min-depth 10 \\\n",
    "  --p-max-depth  160000 \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --o-visualization ../output/alpha-treatment/alpha-rarefaction.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c2a1e0bd-d356-4c77-a56e-3363864f7d86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/25c5090f-2452-4fb6-adee-dd1cde2e2254')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 25c5090f-2452-4fb6-adee-dd1cde2e2254>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/alpha-treatment/alpha-rarefaction.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "efea289d-fe1c-47eb-a549-6179e7655211",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/shared/8TB_HDD_02/stanja/miniconda3/envs/qiime2-2023.5/lib/python3.8/site-packages/sklearn/metrics/pairwise.py:1776: DataConversionWarning: Data was converted to boolean for metric jaccard\n",
      "  warnings.warn(msg, DataConversionWarning)\n",
      "/home/shared/8TB_HDD_02/stanja/miniconda3/envs/qiime2-2023.5/lib/python3.8/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:143: RuntimeWarning: The result contains negative eigenvalues. Please compare their magnitude with the magnitude of some of the largest positive eigenvalues. If the negative ones are smaller, it's probably safe to ignore them, but if they are large in magnitude, the results won't be useful. See the Notes section for more details. The smallest eigenvalue is -0.0006364304980840268 and the largest is 2.5606283832427614.\n",
      "  warn(\n",
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command:\n",
      "\n",
      "faithpd -i /tmp/qiime2/stanja/data/c10eca03-55e7-4350-9e4c-5b42a38eca0e/data/feature-table.biom -t /tmp/qiime2/stanja/data/085b6921-9b93-43d7-83e2-cf3cab988dca/data/tree.nwk -o /tmp/q2-AlphaDiversityFormat-jjv109kp\n",
      "\n",
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command:\n",
      "\n",
      "ssu -i /tmp/qiime2/stanja/data/c10eca03-55e7-4350-9e4c-5b42a38eca0e/data/feature-table.biom -t /tmp/qiime2/stanja/data/085b6921-9b93-43d7-83e2-cf3cab988dca/data/tree.nwk -m unweighted -o /tmp/q2-LSMatFormat-7t61lv50\n",
      "\n",
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command:\n",
      "\n",
      "ssu -i /tmp/qiime2/stanja/data/c10eca03-55e7-4350-9e4c-5b42a38eca0e/data/feature-table.biom -t /tmp/qiime2/stanja/data/085b6921-9b93-43d7-83e2-cf3cab988dca/data/tree.nwk -m weighted_unnormalized -o /tmp/q2-LSMatFormat-1gvllr8f\n",
      "\n",
      "/home/shared/8TB_HDD_02/stanja/miniconda3/envs/qiime2-2023.5/lib/python3.8/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:143: RuntimeWarning: The result contains negative eigenvalues. Please compare their magnitude with the magnitude of some of the largest positive eigenvalues. If the negative ones are smaller, it's probably safe to ignore them, but if they are large in magnitude, the results won't be useful. See the Notes section for more details. The smallest eigenvalue is -0.0015288174756260182 and the largest is 0.19344814121955045.\n",
      "  warn(\n",
      "\u001b[32mSaved FeatureTable[Frequency] to: ../output/diversity-treatment/table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/diversity-treatment/faith-pad-vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/diversity-treatment/observed-features-vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/diversity-treatment/shannon-vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: ../output/diversity-treatment/evenness-vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/diversity-treatment/unweighted-unifrac-distance-matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/diversity-treatment/weighted-unifrac-distance-matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/diversity-treatment/jaccard-distance-matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: ../output/diversity-treatment/bray-curtis-distance-matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/diversity-treatment/unweighted-unifrac-pcoa-results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/diversity-treatment/weighted-unifrac-pcoa-results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/diversity-treatment/jaccard-pcoa-results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: ../output/diversity-treatment/bray-curtis-pcoa-results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/unweighted-unifrac-emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/weighted-unifrac-emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/jaccard-emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/bray-curtis-emperor.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity core-metrics-phylogenetic \\\n",
    "  --i-phylogeny ../output/tree/rooted_tree.qza \\\n",
    "  --i-table ../output/filtered/table-treatment.qza \\\n",
    "  --p-sampling-depth 160000 \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --no-recycle \\\n",
    "  --verbose \\\n",
    "  --o-rarefied-table ../output/diversity-treatment/rarefied-table.qza \\\n",
    "  --o-shannon-vector ../output/diversity-treatment/shannon-vector.qza \\\n",
    "  --o-evenness-vector ../output/diversity-treatment/evenness-vector.qza \\\n",
    "  --o-jaccard-distance-matrix ../output/diversity-treatment/jaccard-distance-matrix.qza \\\n",
    "  --o-bray-curtis-distance-matrix ../output/diversity-treatment/bray-curtis-distance-matrix.qza \\\n",
    "  --o-jaccard-pcoa-results ../output/diversity-treatment/jaccard-pcoa-results.qza \\\n",
    "  --o-bray-curtis-pcoa-results ../output/diversity-treatment/bray-curtis-pcoa-results.qza \\\n",
    "  --o-jaccard-emperor ../output/diversity-treatment/jaccard-emperor.qzv \\\n",
    "  --o-bray-curtis-emperor ../output/diversity-treatment/bray-curtis-emperor.qzv \\\n",
    "  --o-weighted-unifrac-distance-matrix ../output/diversity-treatment/weighted-unifrac-distance-matrix.qza \\\n",
    "  --o-unweighted-unifrac-emperor ../output/diversity-treatment/unweighted-unifrac-emperor.qzv \\\n",
    "  --o-faith-pd-vector ../output/diversity-treatment/faith-pad-vector.qza \\\n",
    "  --o-observed-features-vector ../output/diversity-treatment/observed-features-vector.qza \\\n",
    "  --o-unweighted-unifrac-pcoa-results ../output/diversity-treatment/unweighted-unifrac-pcoa-results.qza \\\n",
    "  --o-weighted-unifrac-pcoa-results ../output/diversity-treatment/weighted-unifrac-pcoa-results.qza \\\n",
    "  --o-unweighted-unifrac-distance-matrix ../output/diversity-treatment/unweighted-unifrac-distance-matrix.qza \\\n",
    "  --o-weighted-unifrac-emperor ../output/diversity-treatment/weighted-unifrac-emperor.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab4f0ad8-959e-4163-a0ed-34f30972dd7f",
   "metadata": {},
   "source": [
    "### Unweighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f57552fb-d87e-46d8-92f1-2f28bcd493d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/6b8ea207-49f8-4a1f-a877-a5a1315ddcfd')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 6b8ea207-49f8-4a1f-a877-a5a1315ddcfd>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/diversity-treatment/unweighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77451e20-6175-4d81-829e-a2391770b99d",
   "metadata": {},
   "source": [
    "### Weighted Unifrac Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1eb02ec8-ef94-48cd-84cc-2346a4a71e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/486b7ff9-9662-4d95-84bf-79b432073272')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 486b7ff9-9662-4d95-84bf-79b432073272>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/diversity-treatment/weighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810aeaf1-0d54-4afd-89dd-a5a9ba1c5a39",
   "metadata": {},
   "source": [
    "### Jaccard Emperor PLot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ff3558a3-558c-4c7e-9548-7c2fc4be2e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/a995e0ce-8435-47fc-8f0d-c60686e8e526')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: a995e0ce-8435-47fc-8f0d-c60686e8e526>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/diversity-treatment/jaccard_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4064ef0f-efe5-48bd-9ebf-8b54817c9edb",
   "metadata": {},
   "source": [
    "### Bray-Curtis Emperor Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "235a6b33-14c3-44a6-bacc-99e053c66e89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/31b2bc93-6fb5-4171-b9f8-4254d52ca1f7')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 31b2bc93-6fb5-4171-b9f8-4254d52ca1f7>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/diversity-treatment/bray_curtis_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce0cac8e-1e40-4dd5-b25e-375dade9b20e",
   "metadata": {},
   "source": [
    "## Alpha Diversity\n",
    "After computing diversity metrics, we can begin to explore the microbial composition of the samples in the context of the sample metadata. This information is present in the sample metadata file `../rawdata/sample-metadata-treatment.tsv`.\n",
    "\n",
    "We’ll first test for associations between categorical metadata columns and alpha diversity data.\n",
    "We’ll do that here for the Faith Phylogenetic Diversity (a measure of community richness):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "820f2e0b-3a03-4192-b666-f1321e8780da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/alpha-treatment/faith-pd-group-significance.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/diversity-treatment/faith_pd_vector.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --o-visualization ../output/alpha-treatment/faith-pd-group-significance.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aacf0ee4-cc0a-4017-9c35-e4409c059127",
   "metadata": {},
   "source": [
    "### Microbial Community Richness: Faith Phylogenetic Diversity\n",
    "\n",
    "##### Which categorical sample metadata columns are most strongly associated with the differences in microbial community richness?\n",
    "##### Are these differences statistically significant?\n",
    "\n",
    "Pae\n",
    "- no significantly different groups\n",
    "\n",
    "Temp\n",
    "- no significantly different groups\n",
    "  \n",
    "PaeTemp\n",
    "- no significantly different groups\n",
    "\n",
    "Colony\n",
    "- no significantly different groups\n",
    "\n",
    "Tank\n",
    "- no significantly different groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7bc2c581-03af-44d8-909d-a685d5370108",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/7c9f52ab-b3dd-4f9b-b920-9659b299c9e3')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 7c9f52ab-b3dd-4f9b-b920-9659b299c9e3>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/alpha-treatment/faith-pd-group-significance.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e5322d40-bde7-471c-9067-0364a9f3cdb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/alpha-treatment/evenness-group-significance.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "  --i-alpha-diversity ../output/diversity-treatment/evenness_vector.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --o-visualization ../output/alpha-treatment/evenness-group-significance.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7088f177-08e6-4fc8-9a53-010742794998",
   "metadata": {},
   "source": [
    "### Microbial Community Evenness, Alpha diversity , Pielou's Evenness\n",
    "\n",
    "##### Which categorical sample metadata columns are most strongly associated with the differences in microbial community evenness? \n",
    "Surprised by Tank result: H1 vs H2, I wouldn't have expected those to be different.\n",
    "\n",
    "##### Are these differences statistically significant?\n",
    "Pae\n",
    "- no significantly different groups\n",
    "- control (n=8) vs peak (n=8), pvalue (0.07)\n",
    "\n",
    "Temp\n",
    "- no significantly different groups\n",
    "- control (n=8) vs peak (n=8), pvalue (0.07)\n",
    "\n",
    "PaeTemp\n",
    "- control-ambient (n=4) vs peak-ambient (n=4), pvalue (0.04)\n",
    "- control-ambient (n=4) vs peak-hot (n=4), pvalue (0.02)\n",
    "  \n",
    "Colony\n",
    "- no categorical data?\n",
    "\n",
    "Tank\n",
    "Kruskal-Wallis pvalue (0.02)\n",
    "- A1 (n=4) vs H1 (n=5), pvalue (0.05)\n",
    "- A2 (n=4) vs H1 (n=5), pvalue (0.01)\n",
    "- H1 (n=5) vs H2 (n=3), pvalue (0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4ed65296-ac10-440b-8819-89d409acc37d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/8472470c-2109-4b46-8f47-c9d76f02230d')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 8472470c-2109-4b46-8f47-c9d76f02230d>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load('../output/alpha-treatment/evenness-group-significance.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f8fc7e8-0742-4e73-b096-7a1b1899f7a3",
   "metadata": {},
   "source": [
    "let's use the Shannon vector in the output directory to create a visualization of alpha diversity across samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "893f1000-6702-4bcf-bd20-4c5d5e84f91f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/alpha-treatment/alpha_groups.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity alpha-group-significance \\\n",
    "    --i-alpha-diversity ../output/diversity-treatment/shannon_vector.qza \\\n",
    "    --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "    --o-visualization ../output/alpha-treatment/alpha_groups.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e1c0068d-9513-4792-a622-1304f8de0885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/9273b438-4972-4536-9c0d-e54924fd7905')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 9273b438-4972-4536-9c0d-e54924fd7905>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/alpha-treatment/alpha_groups.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "123d925c-3c85-406f-834c-0ef6cf32d3b9",
   "metadata": {},
   "source": [
    "## Beta Diversity\n",
    "\n",
    "Next we’ll analyze sample composition in the context of categorical metadata using PERMANOVA (first described in Anderson (2001)) using the beta-group-significance command. The following commands will test whether distances between samples within a group, such as samples from the same body site (e.g., gut), are more similar to each other then they are to samples from the other groups (e.g., tongue, left palm, and right palm). If you call this command with the --p-pairwise parameter, as we’ll do here, it will also perform pairwise tests that will allow you to determine which specific pairs of groups (e.g., tongue and gut) differ from one another, if any. This command can be slow to run, especially when passing --p-pairwise, since it is based on permutation tests. So, unlike the previous commands, we’ll run beta-group-significance on specific columns of metadata that we’re interested in exploring, rather than all metadata columns to which it is applicable. Here we’ll apply this to our unweighted UniFrac distances, using two sample metadata columns, as follows."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f3181c-2a3b-49d0-b4b4-c5f0625cfab6",
   "metadata": {},
   "source": [
    "Let's visualize the beta diversity and see how they separate. For this we'll look at weighted UniFrac. \n",
    "<br>\r\n",
    "\r\n",
    "We can check for 'significant' separation between samples using PERMANOVA. We can do this with the diversity plugin in QIIME2."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6bc837e-63b5-4058-a3da-eed5b79cc1c7",
   "metadata": {},
   "source": [
    "### Adonis Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e47bdb48-e8aa-4e90-a818-3d8a0087db1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/beta-treatment/pae_permanova.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity adonis \\\n",
    "    --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "    --p-formula Pae \\\n",
    "    --p-n-jobs 2 \\\n",
    "    --o-visualization ../output/beta-treatment/pae_permanova.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7088d1a1-5605-454b-8719-07b70f3d8743",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/efef2bae-2f7a-4cf9-88b6-71b3ca989bd1')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: efef2bae-2f7a-4cf9-88b6-71b3ca989bd1>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/pae_permanova.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "4a7e9a52-a978-4f19-ac0d-145853431e36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/beta-treatment/temp_permanova.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity adonis \\\n",
    "    --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "    --p-formula Temp \\\n",
    "    --p-n-jobs 2 \\\n",
    "    --o-visualization ../output/beta-treatment/temp_permanova.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "eb617aff-572b-475c-8efb-414764e98952",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/021f046c-4b65-4321-9b5a-1d9fb43282cf')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 021f046c-4b65-4321-9b5a-1d9fb43282cf>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/temp_permanova.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "70d30816-a111-4167-9cf9-305d9b9b6b3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/beta-treatment/paetemp_permanova.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity adonis \\\n",
    "    --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "    --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "    --p-formula Pae+Temp \\\n",
    "    --p-n-jobs 2 \\\n",
    "    --o-visualization ../output/beta-treatment/paetemp_permanova.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "10d36370-473a-4b38-bb88-b1ef47afffd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/eb39c11c-8703-41da-9fbb-f45481c4ffef')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: eb39c11c-8703-41da-9fbb-f45481c4ffef>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/paetemp_permanova.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dcb1ec7-590d-44cd-8f68-afbbe6f8a841",
   "metadata": {},
   "source": [
    "### Beta-Group Significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b7a10501-6323-4cba-be60-be6e370a7077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/beta-treatment/unweighted-unifrac-pae-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: ../output/beta-treatment/unweighted-unifrac-temp-significance.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: ../output/beta-treatment/unweighted-unifrac-paetemp-significance.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column Pae \\\n",
    "  --o-visualization ../output/beta-treatment/unweighted-unifrac-pae-significance.qzv \\\n",
    "  --p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/beta-treatment/unweighted-unifrac-temp-significance.qzv \\\n",
    "  --p-pairwise\n",
    "\n",
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/unweighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column PaeTemp \\\n",
    "  --o-visualization ../output/beta-treatment/unweighted-unifrac-paetemp-significance.qzv \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2a6a5115-44b4-4c07-bc8f-6088c11926c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/aaefc9fa-92e5-4fc1-a7dd-1116eb0add1c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: aaefc9fa-92e5-4fc1-a7dd-1116eb0add1c>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/unweighted-unifrac-pae-significance.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "746be567-2cf7-4c39-af83-0b9a2ac72a56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/f5ca1b56-467d-4e5c-97fa-71b79e181b6c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: f5ca1b56-467d-4e5c-97fa-71b79e181b6c>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/unweighted-unifrac-temp-significance.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dcb76a8c-49b6-43ae-9532-be45db764c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/c5879810-8b30-45b6-9554-48a43e5752e6')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: c5879810-8b30-45b6-9554-48a43e5752e6>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/beta-treatment/unweighted-unifrac-paetemp-significance.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "306d68f1-5ad8-4879-a77f-f996d4669e06",
   "metadata": {},
   "source": [
    "## PERMDISP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "51835e7f-3cfc-412d-a584-c594768bbd3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/weighted-unifrac-pae-permdisp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column Pae \\\n",
    "  --o-visualization ../output/diversity-treatment/weighted-unifrac-pae-permdisp.qzv \\\n",
    "  --p-method permdisp \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5fb40025-aa7e-4761-afd0-11eb54460c77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/fcfd329a-7f57-4ac0-aca5-769d18d86e3c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: fcfd329a-7f57-4ac0-aca5-769d18d86e3c>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/diversity-treatment/weighted-unifrac-pae-permdisp.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "998172ea-868e-4bc6-9b98-8ce93436ffea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/weighted-unifrac-temp-permdisp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column Temp \\\n",
    "  --o-visualization ../output/diversity-treatment/weighted-unifrac-temp-permdisp.qzv \\\n",
    "  --p-method permdisp \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5ed242bf-f096-419b-b993-f036d7f2a442",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/f573a870-d59e-4710-9dbc-9e87091cfbe6')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: f573a870-d59e-4710-9dbc-9e87091cfbe6>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/diversity-treatment/weighted-unifrac-temp-permdisp.qzv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "d194177f-29cd-41d4-8147-bf845ebd7fff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: ../output/diversity-treatment/weighted-unifrac-paetemp-permdisp.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!qiime diversity beta-group-significance \\\n",
    "  --i-distance-matrix ../output/diversity-treatment/weighted_unifrac_distance_matrix.qza \\\n",
    "  --m-metadata-file ../data/sample-metadata-treatment.tsv \\\n",
    "  --m-metadata-column PaeTemp \\\n",
    "  --o-visualization ../output/diversity-treatment/weighted-unifrac-paetemp-permdisp.qzv \\\n",
    "  --p-method permdisp \\\n",
    "  --p-pairwise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "fff4d083-d20e-4f3d-b047-9cc57b25652f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2/stanja/data/7dd2ba97-bdb0-4d10-8b19-203893867360')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 7dd2ba97-bdb0-4d10-8b19-203893867360>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(\"../output/diversity-treatment/weighted-unifrac-paetemp-permdisp.qzv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e9a30b-96cf-4455-b040-2c21e93fd171",
   "metadata": {},
   "source": [
    "## [Questions to Guide Data Analysis](https://docs.qiime2.org/2023.5/tutorials/atacama-soils/#paired-end-read-analysis-commands:~:text=Questions%20to%20guide%20data%20analysis)\n",
    "What sample metadata or combinations of sample metadata are most strongly associated with the differences in microbial composition of the samples? Are these associations stronger with unweighted UniFrac or with Bray-Curtis? Based on what you know about these metrics, what does that difference suggest? For exploring associations between continuous metadata and sample composition, the commands `qiime metadata distance-matrix` in combination with `qiime diversity mantel` and `qiime diversity bioenv` will be useful. These were not covered in the Moving Pictures tutorial, but you can learn about them by running them with the `--help` parameter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f573e9c-3dee-43c3-8708-e46e0e6075e3",
   "metadata": {},
   "source": [
    "Next Steps:\n",
    "<br>\n",
    "- I want to know which taxa are uniquely present in each treatment, and which ones stay the same\n",
    " - ANCOM, composition plugin (apply ANalysis of Composition of Microbiomes (ANCOM) to identify features that are differentially abundant across groups\n",
    "<br>\n",
    "- Use [PICRUST](https://library.qiime2.org/plugins/q2-picrust2/13/) for functional analysis\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50151432-9f65-4121-b396-b9dcdaf292ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
